<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Map protein alignment gaps to DNA alignment gaps &mdash; PyCogent 1.9 documentation</title>
    
    <link rel="stylesheet" href="../_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.9',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PyCogent 1.9 documentation" href="../index.html" />
    <link rel="up" title="Cogent Usage Examples" href="index.html" />
    <link rel="next" title="Creating and manipulating alignment profiles" href="aln_profile.html" />
    <link rel="prev" title="Getting the reverse complement" href="reverse_complement.html" />     
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRQZ4X-jPDmWcshBrz2j7-fJvuUABRTGWmdiw2G89JpgztGlFGG8hDxRAw"></script>\
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body role="document">
    <div class="header-wrapper" role="banner">
      <div class="header">
        <div class="headertitle"><a
          href="../index.html">PyCogent 1.9 documentation</a></div>
        <div class="rel" role="navigation" aria-label="related navigation">
          <a href="reverse_complement.html" title="Getting the reverse complement"
             accesskey="P">previous</a> |
          <a href="aln_profile.html" title="Creating and manipulating alignment profiles"
             accesskey="N">next</a> |
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="map-protein-alignment-gaps-to-dna-alignment-gaps">
<h1>Map protein alignment gaps to DNA alignment gaps<a class="headerlink" href="#map-protein-alignment-gaps-to-dna-alignment-gaps" title="Permalink to this headline">Â¶</a></h1>
<p><em>Section author: Gavin Huttley</em></p>
<p>Although PyCogent provides a means for directly aligning codon sequences, you may want to use a different approach based on the translate-align-introduce gaps into the original paradigm. After you&#8217;ve translated your codon sequences, and aligned the resulting amino acid sequences, you want to introduce the gaps from the aligned protein sequences back into the original codon sequences. Here&#8217;s how.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="k">import</span> <span class="n">LoadSeqs</span><span class="p">,</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">PROTEIN</span>
</pre></div>
</div>
<p>First I&#8217;m going to construct an artificial example, using the seqs dict as a means to get the data into the Alignment object. The basic idea, however, is that you should already have a set of DNA sequences that are in frame (i.e. position 0 is the 1st codon position), you&#8217;ve translated those sequences and aligned these translated sequences. The result is an alignment of aa sequences and a set of unaligned DNA sequences from which the aa seqs were derived. If your sequences are not in frame you can adjust it by either slicing, or adding N&#8217;s to the beginning of the raw string.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">{</span>
<span class="gp">... </span><span class="s1">&#39;hum&#39;</span><span class="p">:</span> <span class="s1">&#39;AAGCAGATCCAGGAAAGCAGCGAGAATGGCAGCCTGGCCGCGCGCCAGGAGAGGCAGGCCCAGGTCAACCTCACT&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="s1">&#39;mus&#39;</span><span class="p">:</span> <span class="s1">&#39;AAGCAGATCCAGGAGAGCGGCGAGAGCGGCAGCCTGGCCGCGCGGCAGGAGAGGCAGGCCCAAGTCAACCTCACG&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="s1">&#39;rat&#39;</span><span class="p">:</span> <span class="s1">&#39;CTGAACAAGCAGCCACTTTCAAACAAGAAA&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">unaligned_DNA</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">seqs</span><span class="p">,</span> <span class="n">moltype</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">aligned</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">unaligned_DNA</span><span class="o">.</span><span class="n">toFasta</span><span class="p">()</span>
<span class="go">&gt;hum</span>
<span class="go">AAGCAGATCCAGGAAAGCAGCGAGAATGGCAGCCTGGCCGCGCGCCAGGAGAGGCAGGCCCAGGTCAACCTCACT</span>
<span class="go">&gt;mus</span>
<span class="go">AAGCAGATCCAGGAGAGCGGCGAGAGCGGCAGCCTGGCCGCGCGGCAGGAGAGGCAGGCCCAAGTCAACCTCACG</span>
<span class="go">&gt;rat</span>
<span class="go">CTGAACAAGCAGCCACTTTCAAACAAGAAA</span>
</pre></div>
</div>
<p>In order to ensure the alignment algorithm preserves the coding frame, we align the translation of the sequences. We need to translate them first, but note that because the seqs are unaligned they we have to set <code class="docutils literal"><span class="pre">aligned</span> <span class="pre">=</span> <span class="pre">False</span></code>, or we&#8217;ll get an error.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">unaligned_aa</span> <span class="o">=</span> <span class="n">unaligned_DNA</span><span class="o">.</span><span class="n">getTranslation</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">unaligned_aa</span><span class="o">.</span><span class="n">toFasta</span><span class="p">()</span>
<span class="go">&gt;hum</span>
<span class="go">KQIQESSENGSLAARQERQAQVNLT</span>
<span class="go">&gt;mus</span>
<span class="go">KQIQESGESGSLAARQERQAQVNLT</span>
<span class="go">&gt;rat</span>
<span class="go">LNKQPLSNKK</span>
</pre></div>
</div>
<p>The translated seqs can then be written to file, using the method <code class="docutils literal"><span class="pre">writeToFile</span></code>. That file then serves as input for an alignment program. The resulting alignment file can be read back in. (We won&#8217;t write to file in this example.) For this example we will specify the aligned sequences in the dict, rather than from file.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligned_aa_seqs</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;hum&#39;</span><span class="p">:</span> <span class="s1">&#39;KQIQESSENGSLAARQERQAQVNLT&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="s1">&#39;mus&#39;</span><span class="p">:</span> <span class="s1">&#39;KQIQESGESGSLAARQERQAQVNLT&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="s1">&#39;rat&#39;</span><span class="p">:</span> <span class="s1">&#39;LNKQ------PLS---------NKK&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligned_aa</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">aligned_aa_seqs</span><span class="p">,</span> <span class="n">moltype</span> <span class="o">=</span> <span class="n">PROTEIN</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligned_DNA</span> <span class="o">=</span> <span class="n">aligned_aa</span><span class="o">.</span><span class="n">replaceSeqs</span><span class="p">(</span><span class="n">unaligned_DNA</span><span class="p">)</span>
</pre></div>
</div>
<p>Just to be sure, we&#8217;ll check that the DNA sequence has gaps in the right place.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">aligned_DNA</span>
<span class="go">&gt;hum</span>
<span class="go">AAGCAGATCCAGGAAAGCAGCGAGAATGGCAGCCTGGCCGCGCGCCAGGAGAGGCAGGCCCAGGTCAACCTCACT</span>
<span class="go">&gt;rat</span>
<span class="go">CTGAACAAGCAG------------------CCACTTTCA---------------------------AACAAGAAA</span>
<span class="go">&gt;mus</span>
<span class="go">AAGCAGATCCAGGAGAGCGGCGAGAGCGGCAGCCTGGCCGCGCGGCAGGAGAGGCAGGCCCAAGTCAACCTCACG</span>
</pre></div>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
<div class="news">
<table id="feed"><tr><td><h3><a href="http://pycogent.wordpress.com/">PyCogent News and Announcements</a></h3></td>
</tr></table></div>

    
          <h3>Table Of Contents</h3>
          <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Quick installation using pip</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html">The Readme</a></li>
<li class="toctree-l1"><a class="reference internal" href="../coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../data_file_links.html">The data files used in the documentation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Cogent Usage Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cookbook/index.html">PyCogent Cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_notes.html">For Developers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripting_guidelines.html">Scripting guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">Licenses and disclaimer</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ChangeLog.html">Changelog</a></li>
</ul>



          <div role="search">
            <h3 style="margin-top: 1.5em;">Search</h3>
            <form class="search" action="../search.html" method="get">
                <input type="text" name="q" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
            </form>
          </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <div role="navigation" aria-label="related navigaton">
            <a href="reverse_complement.html" title="Getting the reverse complement"
              >previous</a> |
            <a href="aln_profile.html" title="Creating and manipulating alignment profiles"
              >next</a> |
            <a href="../genindex.html" title="General Index"
              >index</a>
          </div>
          <div role="note" aria-label="source link">
              <br/>
              <a href="../_sources/examples/align_codons_to_protein.txt"
                rel="nofollow">Show Source</a>
          </div>
        </div>

        <div class="right">
          
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, PyCogent Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>