<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Using alignment application controllers to align unaligned sequences &mdash; PyCogent 1.9 documentation</title>
    
    <link rel="stylesheet" href="../_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.9',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PyCogent 1.9 documentation" href="../index.html" />
    <link rel="up" title="Cogent Usage Examples" href="index.html" />
    <link rel="next" title="Using phylogeny application controllers to construct phylogenetic trees from alignments" href="phylogeny_app_controllers.html" />
    <link rel="prev" title="Building and using a new application controller" href="building_and_using_an_application_controller.html" />     
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRQZ4X-jPDmWcshBrz2j7-fJvuUABRTGWmdiw2G89JpgztGlFGG8hDxRAw"></script>\
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body role="document">
    <div class="header-wrapper" role="banner">
      <div class="header">
        <div class="headertitle"><a
          href="../index.html">PyCogent 1.9 documentation</a></div>
        <div class="rel" role="navigation" aria-label="related navigation">
          <a href="building_and_using_an_application_controller.html" title="Building and using a new application controller"
             accesskey="P">previous</a> |
          <a href="phylogeny_app_controllers.html" title="Using phylogeny application controllers to construct phylogenetic trees from alignments"
             accesskey="N">next</a> |
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="using-alignment-application-controllers-to-align-unaligned-sequences">
<span id="alignment-controllers"></span><h1>Using alignment application controllers to align unaligned sequences<a class="headerlink" href="#using-alignment-application-controllers-to-align-unaligned-sequences" title="Permalink to this headline">Â¶</a></h1>
<p><em>Section author: Daniel McDonald</em></p>
<p>This document provides examples of how to align sequences using the alignment application controllers. Each alignment application controller module provides the support method <code class="docutils literal"><span class="pre">align_unaligned_seqs</span></code>. This method takes as input a <code class="docutils literal"><span class="pre">SequenceCollection</span></code> object or a dict mapping sequence ids to sequences, the <code class="docutils literal"><span class="pre">MolType</span></code> of the sequences, and an option dict containing specific parameter settings. As output, the method returns an <code class="docutils literal"><span class="pre">Alignment</span></code> object.</p>
<p>First, lets import all of the <code class="docutils literal"><span class="pre">align_unaligned_seqs</span></code> methods:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.app.clustalw</span> <span class="k">import</span> <span class="n">align_unaligned_seqs</span> <span class="k">as</span> <span class="n">clustalw_align_unaligned_seqs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.app.muscle</span> <span class="k">import</span> <span class="n">align_unaligned_seqs</span> <span class="k">as</span> <span class="n">muscle_align_unaligned_seqs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.app.mafft</span> <span class="k">import</span> <span class="n">align_unaligned_seqs</span> <span class="k">as</span> <span class="n">mafft_align_unaligned_seqs</span>
</pre></div>
</div>
<p>Next, we&#8217;ll load our test data. We will be using DNA sequences for this example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.core.moltype</span> <span class="k">import</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="k">import</span> <span class="n">LoadSeqs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">unaligned_seqs</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s1">&#39;data/test2.fasta&#39;</span><span class="p">,</span> <span class="n">aligned</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>Lets align some sequences using default parameters!</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Output is truncated for document formatting</p>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">clustalw_aln</span> <span class="o">=</span> <span class="n">clustalw_align_unaligned_seqs</span><span class="p">(</span><span class="n">unaligned_seqs</span><span class="p">,</span> <span class="n">DNA</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">muscle_aln</span> <span class="o">=</span> <span class="n">muscle_align_unaligned_seqs</span><span class="p">(</span><span class="n">unaligned_seqs</span><span class="p">,</span> <span class="n">DNA</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mafft_aln</span> <span class="o">=</span> <span class="n">mafft_align_unaligned_seqs</span><span class="p">(</span><span class="n">unaligned_seqs</span><span class="p">,</span> <span class="n">DNA</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">clustalw_aln</span>
<span class="go">5 x 60 dna alignment: NineBande[------CGCCA...], Mouse[GCAGTGAGCCA...], ...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">muscle_aln</span>
<span class="go">5 x 60 dna alignment: NineBande[------CGCCA...], Mouse[GCAGTGAGCCA...], ...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mafft_aln</span>
<span class="go">5 x 60 dna alignment: NineBande[------CGCCA...], Mouse[GCAGTGAGCCA...], ...</span>
</pre></div>
</div>
<p>To change specific parameters, simply specify the parameters in a dict and pass it in:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Output is truncated for document formatting</p>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">clustalw_params</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;-gapopen&#39;</span><span class="p">:</span><span class="o">-</span><span class="mi">3</span><span class="p">,</span> <span class="s1">&#39;-quicktree&#39;</span><span class="p">:</span><span class="kc">True</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">clustalw_aln</span> <span class="o">=</span> <span class="n">clustalw_align_unaligned_seqs</span><span class="p">(</span><span class="n">unaligned_seqs</span><span class="p">,</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="n">clustalw_params</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">clustalw_aln</span>
<span class="go">5 x 60 dna alignment: NineBande[------CGCCA...], Mouse[GCAGTGAGCCA...], ...</span>
</pre></div>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
<div class="news">
<table id="feed"><tr><td><h3><a href="http://pycogent.wordpress.com/">PyCogent News and Announcements</a></h3></td>
</tr></table></div>

    
          <h3>Table Of Contents</h3>
          <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Quick installation using pip</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html">The Readme</a></li>
<li class="toctree-l1"><a class="reference internal" href="../coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../data_file_links.html">The data files used in the documentation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Cogent Usage Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cookbook/index.html">PyCogent Cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_notes.html">For Developers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripting_guidelines.html">Scripting guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">Licenses and disclaimer</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ChangeLog.html">Changelog</a></li>
</ul>



          <div role="search">
            <h3 style="margin-top: 1.5em;">Search</h3>
            <form class="search" action="../search.html" method="get">
                <input type="text" name="q" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
            </form>
          </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <div role="navigation" aria-label="related navigaton">
            <a href="building_and_using_an_application_controller.html" title="Building and using a new application controller"
              >previous</a> |
            <a href="phylogeny_app_controllers.html" title="Using phylogeny application controllers to construct phylogenetic trees from alignments"
              >next</a> |
            <a href="../genindex.html" title="General Index"
              >index</a>
          </div>
          <div role="note" aria-label="source link">
              <br/>
              <a href="../_sources/examples/alignment_app_controllers.txt"
                rel="nofollow">Show Source</a>
          </div>
        </div>

        <div class="right">
          
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, PyCogent Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>