

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Use an empirical protein substitution model &mdash; PyCogent 1.5.3 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.5.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PyCogent 1.5.3 documentation" href="../index.html" />
    <link rel="up" title="Cogent Usage Examples" href="index.html" />
    <link rel="next" title="Likelihood analysis of multiple loci" href="testing_multi_loci.html" />
    <link rel="prev" title="Using codon models" href="codon_models.html" />     
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRQZ4X-jPDmWcshBrz2j7-fJvuUABRTGWmdiw2G89JpgztGlFGG8hDxRAw"></script>\
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="testing_multi_loci.html" title="Likelihood analysis of multiple loci"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="codon_models.html" title="Using codon models"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">PyCogent 1.5.3 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Cogent Usage Examples</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="use-an-empirical-protein-substitution-model">
<h1>Use an empirical protein substitution model<a class="headerlink" href="#use-an-empirical-protein-substitution-model" title="Permalink to this headline">Â¶</a></h1>
<p><em>Section author: Gavin Huttley</em></p>
<p>This file contains an example of importing an empirically determined protein substitution matrix such as Dayhoff et al 1978 and using it to create a substitution model. The globin alignment is from the PAML distribution.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span><span class="p">,</span> <span class="n">LoadTree</span><span class="p">,</span> <span class="n">PROTEIN</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.evolve.substitution_model</span> <span class="kn">import</span> <span class="n">EmpiricalProteinMatrix</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.parse.paml_matrix</span> <span class="kn">import</span> <span class="n">PamlMatrixParser</span>
</pre></div>
</div>
<p>Make a tree object.  In this case from a string.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">treestring</span><span class="o">=</span><span class="s">&quot;(((rabbit,rat),human),goat-cow,marsupial);&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">=</span> <span class="n">LoadTree</span><span class="p">(</span><span class="n">treestring</span><span class="o">=</span><span class="n">treestring</span><span class="p">)</span>
</pre></div>
</div>
<p>Import the alignment, explicitly setting the <tt class="docutils literal"><span class="pre">moltype</span></tt> to be protein</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">al</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="s">&#39;data/abglobin_aa.phylip&#39;</span><span class="p">,</span>
<span class="gp">... </span>               <span class="n">interleaved</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
<span class="gp">... </span>               <span class="n">moltype</span><span class="o">=</span><span class="n">PROTEIN</span><span class="p">,</span>
<span class="gp">... </span>               <span class="p">)</span>
</pre></div>
</div>
<p>Open the file that contains the empirical matrix and parse the matrix and frequencies.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">matrix_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;data/dayhoff.dat&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The <tt class="docutils literal"><span class="pre">PamlMatrixParser</span></tt> will import the matrix and frequency from files designed for Yang&#8217;s PAML package.  This format is the lower half of the matrix in three letter amino acid name order white space delineated followed by motif frequencies in the same order.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">empirical_matrix</span><span class="p">,</span> <span class="n">empirical_frequencies</span> <span class="o">=</span> <span class="n">PamlMatrixParser</span><span class="p">(</span><span class="n">matrix_file</span><span class="p">)</span>
</pre></div>
</div>
<p>Create an Empirical Protein Matrix Substitution model object.  This will take the unscaled empirical matrix and use it and the motif frequencies to create a scaled Q matrix.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span> <span class="o">=</span> <span class="n">EmpiricalProteinMatrix</span><span class="p">(</span><span class="n">empirical_matrix</span><span class="p">,</span> <span class="n">empirical_frequencies</span><span class="p">)</span>
</pre></div>
</div>
<p>Make a parameter controller, likelihood function object and optimise.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">makeLikelihoodFunction</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setAlignment</span><span class="p">(</span><span class="n">al</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span><span class="o">.</span><span class="n">getLogLikelihood</span><span class="p">()</span>
<span class="go">-1706...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span>
<span class="go">Likelihood Function Table</span>
<span class="go">=============================</span>
<span class="go">     edge    parent    length</span>
<span class="go">-----------------------------</span>
<span class="go">   rabbit    edge.0    0.0785</span>
<span class="go">      rat    edge.0    0.1750</span>
<span class="go">   edge.0    edge.1    0.0324</span>
<span class="go">    human    edge.1    0.0545</span>
<span class="go">   edge.1      root    0.0269</span>
<span class="go"> goat-cow      root    0.0972</span>
<span class="go">marsupial      root    0.2424</span>
<span class="go">-----------------------------</span>
<span class="go">===============</span>
<span class="go">motif    mprobs</span>
<span class="go">---------------</span>
<span class="go">    A    0.0871</span>
<span class="go">    C    0.0335</span>
<span class="go">    D    0.0469</span>
<span class="go">    E    0.0495</span>
<span class="go">    F    0.0398</span>
<span class="go">    G    0.0886</span>
<span class="go">    H    0.0336</span>
<span class="go">    I    0.0369</span>
<span class="go">    K    0.0805</span>
<span class="go">    L    0.0854</span>
<span class="go">    M    0.0148</span>
<span class="go">    N    0.0404</span>
<span class="go">    P    0.0507</span>
<span class="go">    Q    0.0383</span>
<span class="go">    R    0.0409</span>
<span class="go">    S    0.0696</span>
<span class="go">    T    0.0585</span>
<span class="go">    V    0.0647</span>
<span class="go">    W    0.0105</span>
<span class="go">    Y    0.0299</span>
<span class="go">---------------</span>
</pre></div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div class="news">
<table id="feed"><tr><td><h3><a href="http://pycogent.wordpress.com/">PyCogent News and Announcements</a></h3></td>
</tr></table></div>

    


  <h4>Previous topic</h4>
  <p class="topless"><a href="codon_models.html"
                        title="previous chapter">Using codon models</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="testing_multi_loci.html"
                        title="next chapter">Likelihood analysis of multiple loci</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/examples/empirical_protein_models.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="testing_multi_loci.html" title="Likelihood analysis of multiple loci"
             >next</a> |</li>
        <li class="right" >
          <a href="codon_models.html" title="Using codon models"
             >previous</a> |</li>
        <li><a href="../index.html">PyCogent 1.5.3 documentation</a> &raquo;</li>
          <li><a href="index.html" >Cogent Usage Examples</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2009, PyCogent Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>