

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Evaluate process heterogeneity using a Hidden Markov Model &mdash; PyCogent 1.5.3 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.5.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PyCogent 1.5.3 documentation" href="../index.html" />
    <link rel="up" title="Cogent Usage Examples" href="index.html" />
    <link rel="next" title="Seqsim Simple Alignment Simulation Example" href="seqsim_alignment_simulation.html" />
    <link rel="prev" title="Analysis of rate heterogeneity" href="rate_heterogeneity.html" />     
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRQZ4X-jPDmWcshBrz2j7-fJvuUABRTGWmdiw2G89JpgztGlFGG8hDxRAw"></script>\
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="seqsim_alignment_simulation.html" title="Seqsim Simple Alignment Simulation Example"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="rate_heterogeneity.html" title="Analysis of rate heterogeneity"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">PyCogent 1.5.3 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Cogent Usage Examples</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="evaluate-process-heterogeneity-using-a-hidden-markov-model">
<span id="rate-heterogeneity-hmm"></span><h1>Evaluate process heterogeneity using a Hidden Markov Model<a class="headerlink" href="#evaluate-process-heterogeneity-using-a-hidden-markov-model" title="Permalink to this headline">¶</a></h1>
<p><em>Section author: Gavin Huttley</em></p>
<p>The existence of rate heterogeneity in the evolution of biological sequences is well known. Typically such an evolutionary property is evaluated using so-called site-heterogeneity models. These models postulate the existence of discrete classes of sites, where sites within a class evolve according to a specific rate that is distinct from the rates of the other classes. These models retain the assumption that alignment columns evolve independently. One can naturally ask the question of whether rate classes occur randomly throughout the sequence or whether they are in fact auto-correlated - meaning sites of a class tend to cluster together. Because we do not have, <em>a priori</em>, a basis for classifying the sites the models are specified such that each column can belong to any of the designated site classes and the likelihood is computed across all possible classifications. Post numerical optimisation we can calculate the posterior probability a site column belongs to a specific site class. In <tt class="docutils literal"><span class="pre">cogent</span></tt>, site classes are referred to as <tt class="docutils literal"><span class="pre">bins</span></tt> and so we refer to bin probabilities etc ...</p>
<p>To illustrate how to evaluate these hypotheses formally we specify 3 nested hypotheses: (i) Ho: no rate heterogeneity; (ii) Ha(1): two classes of sites - fast and slow, but independent sites; (iii) Ha(2): fast and slowly evolving sites are auto-correlated (meaning a sites class is correlated with that of its&#8217; immediate neighbours).</p>
<p>It is also possible to apply these models to different types of changes and we illustrate this with a single parameterisation at the end.</p>
<p>First import standard components necessary for all of the following calculations. As the likelihood ratio tests (LRT) involve nested hypotheses we will employ the chi-square approximation for assessing statistical significance.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.evolve.substitution_model</span> <span class="kn">import</span> <span class="n">Nucleotide</span><span class="p">,</span> <span class="n">predicate</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span><span class="p">,</span> <span class="n">LoadTree</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.maths.stats</span> <span class="kn">import</span> <span class="n">chisqprob</span>
</pre></div>
</div>
<p>Load the alignment and tree.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="s">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">LoadTree</span><span class="p">(</span><span class="s">&quot;data/test.tree&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="section" id="model-ho-no-rate-heterogeneity">
<h2>Model Ho: no rate heterogeneity<a class="headerlink" href="#model-ho-no-rate-heterogeneity" title="Permalink to this headline">¶</a></h2>
<p>We define a HKY model of nucleotide substitution, which has a transition parameter. This is defined using the <tt class="docutils literal"><span class="pre">MotifChange</span></tt> class, by specifying a transition as <strong>not</strong> a transversion (<tt class="docutils literal"><span class="pre">~MotifChange('R','Y')</span></tt>).</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">MotifChange</span> <span class="o">=</span> <span class="n">predicate</span><span class="o">.</span><span class="n">MotifChange</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">treat_gap</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">recode_gaps</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">model_gaps</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">kappa</span> <span class="o">=</span> <span class="p">(</span><span class="o">~</span><span class="n">MotifChange</span><span class="p">(</span><span class="s">&#39;R&#39;</span><span class="p">,</span> <span class="s">&#39;Y&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">aliased</span><span class="p">(</span><span class="s">&#39;kappa&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span> <span class="o">=</span> <span class="n">Nucleotide</span><span class="p">(</span><span class="n">predicates</span><span class="o">=</span><span class="p">[</span><span class="n">kappa</span><span class="p">],</span> <span class="o">**</span><span class="n">treat_gap</span><span class="p">)</span>
</pre></div>
</div>
<p>We specify a null model with no bins, and optimise it.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf_one</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">makeLikelihoodFunction</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">digits</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">space</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf_one</span><span class="o">.</span><span class="n">setAlignment</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf_one</span><span class="o">.</span><span class="n">optimise</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lnL_one</span> <span class="o">=</span> <span class="n">lf_one</span><span class="o">.</span><span class="n">getLogLikelihood</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df_one</span> <span class="o">=</span> <span class="n">lf_one</span><span class="o">.</span><span class="n">getNumFreeParams</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf_one</span>
<span class="go">Likelihood Function Table</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 4.10</span>
<span class="go">-----</span>
<span class="go">===========================</span>
<span class="go">     edge   parent   length</span>
<span class="go">---------------------------</span>
<span class="go">    Human   edge.0     0.03</span>
<span class="go">HowlerMon   edge.0     0.04</span>
<span class="go">   edge.0   edge.1     0.04</span>
<span class="go">    Mouse   edge.1     0.28</span>
<span class="go">   edge.1     root     0.02</span>
<span class="go">NineBande     root     0.09</span>
<span class="go"> DogFaced     root     0.11</span>
<span class="go">---------------------------</span>
<span class="go">==============</span>
<span class="go">motif   mprobs</span>
<span class="go">--------------</span>
<span class="go">    T     0.23</span>
<span class="go">    C     0.19</span>
<span class="go">    A     0.37</span>
<span class="go">    G     0.21</span>
<span class="go">--------------</span>
</pre></div>
</div>
</div>
<div class="section" id="model-ha-1-two-classes-of-gamma-distributed-but-independent-sites">
<h2>Model Ha(1): two classes of gamma distributed but independent sites<a class="headerlink" href="#model-ha-1-two-classes-of-gamma-distributed-but-independent-sites" title="Permalink to this headline">¶</a></h2>
<p>Our next hypothesis is that there are two rate classes, or bins, with rates gamma distributed. We will restrict the bin probabilities to be equal.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">bin_submod</span> <span class="o">=</span> <span class="n">Nucleotide</span><span class="p">(</span><span class="n">predicates</span><span class="o">=</span><span class="p">[</span><span class="n">kappa</span><span class="p">],</span> <span class="n">ordered_param</span><span class="o">=</span><span class="s">&#39;rate&#39;</span><span class="p">,</span>
<span class="gp">... </span>                     <span class="n">distribution</span><span class="o">=</span><span class="s">&#39;gamma&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">treat_gap</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf_bins</span> <span class="o">=</span> <span class="n">bin_submod</span><span class="o">.</span><span class="n">makeLikelihoodFunction</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="gp">... </span>                            <span class="n">sites_independent</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">digits</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">space</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf_bins</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;bprobs&#39;</span><span class="p">,</span> <span class="n">is_constant</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf_bins</span><span class="o">.</span><span class="n">setAlignment</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf_bins</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lnL_bins</span> <span class="o">=</span> <span class="n">lf_bins</span><span class="o">.</span><span class="n">getLogLikelihood</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df_bins</span> <span class="o">=</span> <span class="n">lf_bins</span><span class="o">.</span><span class="n">getNumFreeParams</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="n">df_bins</span> <span class="o">==</span> <span class="mi">9</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf_bins</span>
<span class="go">Likelihood Function Table</span>
<span class="go">==================</span>
<span class="go">kappa   rate_shape</span>
<span class="go">------------------</span>
<span class="go"> 4.38         1.26</span>
<span class="go">------------------</span>
<span class="go">===========================</span>
<span class="go">     edge   parent   length</span>
<span class="go">---------------------------</span>
<span class="go">    Human   edge.0     0.03</span>
<span class="go">HowlerMon   edge.0     0.04</span>
<span class="go">   edge.0   edge.1     0.04</span>
<span class="go">    Mouse   edge.1     0.31</span>
<span class="go">   edge.1     root     0.02</span>
<span class="go">NineBande     root     0.10</span>
<span class="go"> DogFaced     root     0.12</span>
<span class="go">---------------------------</span>
<span class="go">====================</span>
<span class="go"> bin   bprobs   rate</span>
<span class="go">--------------------</span>
<span class="go">bin0     0.50   0.41</span>
<span class="go">bin1     0.50   1.59</span>
<span class="go">--------------------</span>
<span class="go">==============</span>
<span class="go">motif   mprobs</span>
<span class="go">--------------</span>
<span class="go">    T     0.23</span>
<span class="go">    C     0.19</span>
<span class="go">    A     0.37</span>
<span class="go">    G     0.21</span>
<span class="go">--------------</span>
</pre></div>
</div>
</div>
<div class="section" id="model-ha-2-fast-and-slowly-evolving-sites-are-auto-correlated">
<h2>Model Ha(2): fast and slowly evolving sites are auto-correlated<a class="headerlink" href="#model-ha-2-fast-and-slowly-evolving-sites-are-auto-correlated" title="Permalink to this headline">¶</a></h2>
<p>We then specify a model with switches for changing between site-classes, the HMM part. The setup is almost identical to that for above with the sole difference being setting the <tt class="docutils literal"><span class="pre">sites_independent=False</span></tt>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf_patches</span> <span class="o">=</span> <span class="n">bin_submod</span><span class="o">.</span><span class="n">makeLikelihoodFunction</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="gp">... </span>                        <span class="n">sites_independent</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">digits</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">space</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf_patches</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;bprobs&#39;</span><span class="p">,</span> <span class="n">is_constant</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf_patches</span><span class="o">.</span><span class="n">setAlignment</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf_patches</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lnL_patches</span> <span class="o">=</span> <span class="n">lf_patches</span><span class="o">.</span><span class="n">getLogLikelihood</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df_patches</span> <span class="o">=</span> <span class="n">lf_patches</span><span class="o">.</span><span class="n">getNumFreeParams</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf_patches</span>
<span class="go">Likelihood Function Table</span>
<span class="go">===============================</span>
<span class="go">bin_switch   kappa   rate_shape</span>
<span class="go">-------------------------------</span>
<span class="go">      0.56    4.42         1.16</span>
<span class="go">-------------------------------</span>
<span class="go">===========================</span>
<span class="go">     edge   parent   length</span>
<span class="go">---------------------------</span>
<span class="go">    Human   edge.0     0.03</span>
<span class="go">HowlerMon   edge.0     0.04</span>
<span class="go">   edge.0   edge.1     0.04</span>
<span class="go">    Mouse   edge.1     0.31</span>
<span class="go">   edge.1     root     0.02</span>
<span class="go">NineBande     root     0.10</span>
<span class="go"> DogFaced     root     0.12</span>
<span class="go">---------------------------</span>
<span class="go">====================</span>
<span class="go"> bin   bprobs   rate</span>
<span class="go">--------------------</span>
<span class="go">bin0     0.50   0.39</span>
<span class="go">bin1     0.50   1.61</span>
<span class="go">--------------------</span>
<span class="go">==============</span>
<span class="go">motif   mprobs</span>
<span class="go">--------------</span>
<span class="go">    T     0.23</span>
<span class="go">    C     0.19</span>
<span class="go">    A     0.37</span>
<span class="go">    G     0.21</span>
<span class="go">--------------</span>
</pre></div>
</div>
<p>We use the following short function to compute the LR test statistic.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">LR</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">alt</span><span class="p">,</span> <span class="n">null</span><span class="p">:</span> <span class="mi">2</span> <span class="o">*</span> <span class="p">(</span><span class="n">alt</span> <span class="o">-</span> <span class="n">null</span><span class="p">)</span>
</pre></div>
</div>
<p>We conduct the test between the sequentially nested models.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lr</span> <span class="o">=</span> <span class="n">LR</span><span class="p">(</span><span class="n">lnL_bins</span><span class="p">,</span> <span class="n">lnL_one</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lr</span>
<span class="go">22...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.4f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">chisqprob</span><span class="p">(</span><span class="n">lr</span><span class="p">,</span> <span class="n">df_patches</span><span class="o">-</span><span class="n">df_bins</span><span class="p">)</span>
<span class="go">0.0000</span>
</pre></div>
</div>
<p>The stationary bin probabilities are labelled as <tt class="docutils literal"><span class="pre">bprobs</span></tt> and can be obtained as follows.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">bprobs</span> <span class="o">=</span> <span class="n">lf_patches</span><span class="o">.</span><span class="n">getParamValue</span><span class="p">(</span><span class="s">&#39;bprobs&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.1f</span><span class="s"> : </span><span class="si">%.1f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">bprobs</span><span class="p">)</span>
<span class="go">0.5 : 0.5</span>
</pre></div>
</div>
<p>Of greater interest here (given the model was set up so the bin probabilities were equal, i.e. <tt class="docutils literal"><span class="pre">is_constant=True</span></tt>) are the posterior probabilities as those allow classification of sites. The result is a <tt class="docutils literal"><span class="pre">DictArray</span></tt> class instance, which behaves like a dictionary.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">pp</span> <span class="o">=</span> <span class="n">lf_patches</span><span class="o">.</span><span class="n">getBinProbs</span><span class="p">()</span>
</pre></div>
</div>
<p>If we want to know the posterior probability the 21st position belongs to <tt class="docutils literal"><span class="pre">bin0</span></tt>, we can determine it as:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">pp</span><span class="p">[</span><span class="s">&#39;bin0&#39;</span><span class="p">][</span><span class="mi">20</span><span class="p">]</span>
<span class="go">0.8...</span>
</pre></div>
</div>
</div>
<div class="section" id="a-model-with-patches-of-kappa">
<h2>A model with patches of <tt class="docutils literal"><span class="pre">kappa</span></tt><a class="headerlink" href="#a-model-with-patches-of-kappa" title="Permalink to this headline">¶</a></h2>
<p>In this example we model sequence evolution where there are 2 classes of sites distinguished by their <tt class="docutils literal"><span class="pre">kappa</span></tt> parameters. We need to know what value of <tt class="docutils literal"><span class="pre">kappa</span></tt> to specify the delineation of the bin boundaries. We can determine this from the null model (<tt class="docutils literal"><span class="pre">lf_one</span></tt>). For this use case, we also need to use a <tt class="docutils literal"><span class="pre">numpy.array</span></tt>, so we&#8217;ll import that.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">numpy</span> <span class="kn">import</span> <span class="n">array</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">single_kappa</span> <span class="o">=</span> <span class="n">lf_one</span><span class="o">.</span><span class="n">getParamValue</span><span class="p">(</span><span class="s">&#39;kappa&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>We then construct the substitution model in a different way to that when evaluating generic rate heterogeneity (above).</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">kappa_bin_submod</span> <span class="o">=</span> <span class="n">Nucleotide</span><span class="p">(</span><span class="n">predicates</span><span class="o">=</span><span class="p">[</span><span class="n">kappa</span><span class="p">],</span> <span class="o">**</span><span class="n">treat_gap</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf_kappa</span> <span class="o">=</span> <span class="n">kappa_bin_submod</span><span class="o">.</span><span class="n">makeLikelihoodFunction</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span>
<span class="gp">... </span>     <span class="n">bins</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;slow&#39;</span><span class="p">,</span> <span class="s">&#39;fast&#39;</span><span class="p">],</span> <span class="n">sites_independent</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">digits</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="gp">... </span>     <span class="n">space</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
<p>To improve the likelihood fitting it is desirable to set starting values in the model that result in it&#8217;s initial likelihood being that of the null model (or as close as possible). To do this, we&#8217;re going to define an arbitrarily small value (<tt class="docutils literal"><span class="pre">epsilon</span></tt>) which we use to provide the starting value to the two bins as slightly smaller/greater than <tt class="docutils literal"><span class="pre">single_kappa</span></tt> for the slow/fast bins respectively. At the same time we set the upper/lower bin boundaries.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">epsilon</span> <span class="o">=</span> <span class="mf">1e-6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf_kappa</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="n">kappa</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="n">single_kappa</span><span class="o">-</span><span class="n">epsilon</span><span class="p">,</span>
<span class="gp">... </span>                     <span class="n">upper</span><span class="o">=</span><span class="n">single_kappa</span><span class="p">,</span> <span class="nb">bin</span><span class="o">=</span><span class="s">&#39;slow&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf_kappa</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="n">kappa</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="n">single_kappa</span><span class="o">+</span><span class="n">epsilon</span><span class="p">,</span>
<span class="gp">... </span>                     <span class="n">lower</span><span class="o">=</span><span class="n">single_kappa</span><span class="p">,</span> <span class="nb">bin</span><span class="o">=</span><span class="s">&#39;fast&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>We then illustrate how to adjust the bin probabilities, here doing it so that one of them is nearly 1, the other nearly 0. This ensures the likelihood will be near identical to that of <tt class="docutils literal"><span class="pre">lf_one</span></tt> and as a result the optimisation step will actually improve fit over the simpler model.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf_kappa</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;bprobs&#39;</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">init</span><span class="o">=</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="o">-</span><span class="n">epsilon</span><span class="p">,</span> <span class="mf">0.0</span><span class="o">+</span><span class="n">epsilon</span><span class="p">]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf_kappa</span><span class="o">.</span><span class="n">setAlignment</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf_kappa</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf_kappa</span>
<span class="go">Likelihood Function Table</span>
<span class="go">==========</span>
<span class="go">bin_switch</span>
<span class="go">----------</span>
<span class="go">       0.6</span>
<span class="go">----------</span>
<span class="go">=====================</span>
<span class="go"> bin   bprobs   kappa</span>
<span class="go">---------------------</span>
<span class="go">slow      0.8     3.0</span>
<span class="go">fast      0.2    23.3</span>
<span class="go">---------------------</span>
<span class="go">===========================</span>
<span class="go">     edge   parent   length</span>
<span class="go">---------------------------</span>
<span class="go">    Human   edge.0      0.0</span>
<span class="go">HowlerMon   edge.0      0.0</span>
<span class="go">   edge.0   edge.1      0.0</span>
<span class="go">    Mouse   edge.1      0.3</span>
<span class="go">   edge.1     root      0.0</span>
<span class="go">NineBande     root      0.1</span>
<span class="go"> DogFaced     root      0.1</span>
<span class="go">---------------------------</span>
<span class="go">==============</span>
<span class="go">motif   mprobs</span>
<span class="go">--------------</span>
<span class="go">    T      0.2</span>
<span class="go">    C      0.2</span>
<span class="go">    A      0.4</span>
<span class="go">    G      0.2</span>
<span class="go">--------------</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf_kappa</span><span class="o">.</span><span class="n">getLogLikelihood</span><span class="p">()</span>
<span class="go">-8749.3...</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div class="news">
<table id="feed"><tr><td><h3><a href="http://pycogent.wordpress.com/">PyCogent News and Announcements</a></h3></td>
</tr></table></div>

    
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Evaluate process heterogeneity using a Hidden Markov Model</a><ul>
<li><a class="reference internal" href="#model-ho-no-rate-heterogeneity">Model Ho: no rate heterogeneity</a></li>
<li><a class="reference internal" href="#model-ha-1-two-classes-of-gamma-distributed-but-independent-sites">Model Ha(1): two classes of gamma distributed but independent sites</a></li>
<li><a class="reference internal" href="#model-ha-2-fast-and-slowly-evolving-sites-are-auto-correlated">Model Ha(2): fast and slowly evolving sites are auto-correlated</a></li>
<li><a class="reference internal" href="#a-model-with-patches-of-kappa">A model with patches of <tt class="docutils literal"><span class="pre">kappa</span></tt></a></li>
</ul>
</li>
</ul>



  <h4>Previous topic</h4>
  <p class="topless"><a href="rate_heterogeneity.html"
                        title="previous chapter">Analysis of rate heterogeneity</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="seqsim_alignment_simulation.html"
                        title="next chapter">Seqsim Simple Alignment Simulation Example</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/examples/hmm_par_heterogeneity.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="seqsim_alignment_simulation.html" title="Seqsim Simple Alignment Simulation Example"
             >next</a> |</li>
        <li class="right" >
          <a href="rate_heterogeneity.html" title="Analysis of rate heterogeneity"
             >previous</a> |</li>
        <li><a href="../index.html">PyCogent 1.5.3 documentation</a> &raquo;</li>
          <li><a href="index.html" >Cogent Usage Examples</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2009, PyCogent Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>