<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Making a phylogenetic tree from a protein sequence alignment &mdash; PyCogent 1.9 documentation</title>
    
    <link rel="stylesheet" href="../_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.9',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PyCogent 1.9 documentation" href="../index.html" />
    <link rel="up" title="Cogent Usage Examples" href="index.html" />
    <link rel="next" title="PyCogent Cookbook" href="../cookbook/index.html" />
    <link rel="prev" title="Phylogenetic reconstruction by least squares" href="phylo_by_ls.html" />     
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRQZ4X-jPDmWcshBrz2j7-fJvuUABRTGWmdiw2G89JpgztGlFGG8hDxRAw"></script>\
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body role="document">
    <div class="header-wrapper" role="banner">
      <div class="header">
        <div class="headertitle"><a
          href="../index.html">PyCogent 1.9 documentation</a></div>
        <div class="rel" role="navigation" aria-label="related navigation">
          <a href="phylo_by_ls.html" title="Phylogenetic reconstruction by least squares"
             accesskey="P">previous</a> |
          <a href="../cookbook/index.html" title="PyCogent Cookbook"
             accesskey="N">next</a> |
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="making-a-phylogenetic-tree-from-a-protein-sequence-alignment">
<h1>Making a phylogenetic tree from a protein sequence alignment<a class="headerlink" href="#making-a-phylogenetic-tree-from-a-protein-sequence-alignment" title="Permalink to this headline">Â¶</a></h1>
<p><em>Section author: Gavin Huttley</em></p>
<p>In this example we pull together the distance calculation and tree building with the additional twist of using an empirical protein substitution matrix. We will therefore be computing the tree from a protein sequence alignment. We will first do the standard cogent import for <code class="docutils literal"><span class="pre">LoadSeqs</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="k">import</span> <span class="n">LoadSeqs</span><span class="p">,</span> <span class="n">PROTEIN</span>
</pre></div>
</div>
<p>We will use an empirical protein substitution matrix.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.evolve.models</span> <span class="k">import</span> <span class="n">JTT92</span>
</pre></div>
</div>
<p>The next components we need are for computing the matrix of pairwise sequence distances and then for estimating a neighbour joining tree from those distances.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.phylo</span> <span class="k">import</span> <span class="n">nj</span><span class="p">,</span> <span class="n">distance</span>
</pre></div>
</div>
<p>Now load our sequence alignment, explicitly setting the alphabet to be protein.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="s1">&#39;data/abglobin_aa.phylip&#39;</span><span class="p">,</span> <span class="n">interleaved</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>                <span class="n">moltype</span><span class="o">=</span><span class="n">PROTEIN</span><span class="p">)</span>
</pre></div>
</div>
<p>Create an Empirical Protein Matrix Substitution model object. This will take the unscaled empirical matrix and use it and the motif frequencies to create a scaled Q matrix.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span> <span class="o">=</span> <span class="n">JTT92</span><span class="p">()</span>
</pre></div>
</div>
<p>We now use this and the alignment to construct a distance calculator.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">distance</span><span class="o">.</span><span class="n">EstimateDistances</span><span class="p">(</span><span class="n">aln</span><span class="p">,</span> <span class="n">submodel</span> <span class="o">=</span> <span class="n">sm</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>The resulting distances are passed to the nj function.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mytree</span> <span class="o">=</span> <span class="n">nj</span><span class="o">.</span><span class="n">nj</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">getPairwiseDistances</span><span class="p">())</span>
</pre></div>
</div>
<p>The shape of the resulting tree can be readily view by printing <code class="docutils literal"><span class="pre">mytree.asciiArt()</span></code>. The result will be equivalent to.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>          <span class="o">/-</span><span class="n">human</span>
         <span class="o">|</span>
         <span class="o">|</span>          <span class="o">/-</span><span class="n">rabbit</span>
<span class="o">-</span><span class="n">root</span><span class="o">----|-</span><span class="n">edge</span><span class="o">.</span><span class="mi">1</span><span class="o">--|</span>
         <span class="o">|</span>          \<span class="o">-</span><span class="n">rat</span>
         <span class="o">|</span>
         <span class="o">|</span>          <span class="o">/-</span><span class="n">goat</span><span class="o">-</span><span class="n">cow</span>
          \<span class="n">edge</span><span class="o">.</span><span class="mi">0</span><span class="o">--|</span>
                    \<span class="o">-</span><span class="n">marsupial</span>
</pre></div>
</div>
<p>This tree can be saved to file, the <code class="docutils literal"><span class="pre">with_distances</span></code> argument specifies that branch lengths are to be included in the newick formatted output.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mytree</span><span class="o">.</span><span class="n">writeToFile</span><span class="p">(</span><span class="s1">&#39;test_nj.tree&#39;</span><span class="p">,</span> <span class="n">with_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
<div class="news">
<table id="feed"><tr><td><h3><a href="http://pycogent.wordpress.com/">PyCogent News and Announcements</a></h3></td>
</tr></table></div>

    
          <h3>Table Of Contents</h3>
          <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Quick installation using pip</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html">The Readme</a></li>
<li class="toctree-l1"><a class="reference internal" href="../coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../data_file_links.html">The data files used in the documentation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Cogent Usage Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cookbook/index.html">PyCogent Cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_notes.html">For Developers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripting_guidelines.html">Scripting guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">Licenses and disclaimer</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ChangeLog.html">Changelog</a></li>
</ul>



          <div role="search">
            <h3 style="margin-top: 1.5em;">Search</h3>
            <form class="search" action="../search.html" method="get">
                <input type="text" name="q" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
            </form>
          </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <div role="navigation" aria-label="related navigaton">
            <a href="phylo_by_ls.html" title="Phylogenetic reconstruction by least squares"
              >previous</a> |
            <a href="../cookbook/index.html" title="PyCogent Cookbook"
              >next</a> |
            <a href="../genindex.html" title="General Index"
              >index</a>
          </div>
          <div role="note" aria-label="source link">
              <br/>
              <a href="../_sources/examples/maketree_from_proteinseqs.txt"
                rel="nofollow">Show Source</a>
          </div>
        </div>

        <div class="right">
          
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, PyCogent Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>