

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>A test of the neutral theory &mdash; PyCogent 1.5.3 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.5.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PyCogent 1.5.3 documentation" href="../index.html" />
    <link rel="up" title="Cogent Usage Examples" href="index.html" />
    <link rel="next" title="Allowing substitution model parameters to differ between branches" href="scope_model_params_on_trees.html" />
    <link rel="prev" title="Performing a relative rate test" href="relative_rate.html" />     
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRQZ4X-jPDmWcshBrz2j7-fJvuUABRTGWmdiw2G89JpgztGlFGG8hDxRAw"></script>\
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="scope_model_params_on_trees.html" title="Allowing substitution model parameters to differ between branches"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="relative_rate.html" title="Performing a relative rate test"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">PyCogent 1.5.3 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Cogent Usage Examples</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="a-test-of-the-neutral-theory">
<h1>A test of the neutral theory<a class="headerlink" href="#a-test-of-the-neutral-theory" title="Permalink to this headline">Â¶</a></h1>
<p><em>Section author: Gavin Huttley</em></p>
<p>This file contains an example for performing a likelihood ratio test of neutrality. The test compares a model where the codon model parameter omega is constrained to be the same for all edges against one where each edge has its&#8217; own omega. From cogent import all the components we need.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span><span class="p">,</span> <span class="n">LoadTree</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.evolve.models</span> <span class="kn">import</span> <span class="n">MG94GTR</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.maths</span> <span class="kn">import</span> <span class="n">stats</span>
</pre></div>
</div>
<p>Get your alignment and tree.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">al</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="s">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">=</span> <span class="n">LoadTree</span><span class="p">(</span><span class="s">&quot;data/test.tree&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We use a Goldman Yang 1994 model.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span> <span class="o">=</span> <span class="n">MG94GTR</span><span class="p">()</span>
</pre></div>
</div>
<p>Make the controller object</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">makeLikelihoodFunction</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">digits</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">space</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
<p>Get the likelihood function object this object performs the actual likelihood calculation.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setAlignment</span><span class="p">(</span><span class="n">al</span><span class="p">)</span>
</pre></div>
</div>
<p>By default, parameters other than branch lengths are treated as global in scope, so we don&#8217;t need to do anything special here. We can influence how rigorous the optimisation will be, and switch between the global and local optimisers provided in the toolkit using arguments to the optimise method. The <tt class="docutils literal"><span class="pre">global_tolerance=1.0</span></tt> argument specifies conditions for an early break from simulated annealing which will be automatically followed by the Powell local optimiser. .. note:: the &#8216;results&#8217; are of course nonsense.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">global_tolerance</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">)</span>
</pre></div>
</div>
<p>View the resulting maximum-likelihood parameter values</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span>
<span class="go">Likelihood Function Table</span>
<span class="go">===================================</span>
<span class="go"> A/C   A/G   A/T   C/G   C/T  omega</span>
<span class="go">-----------------------------------</span>
<span class="go">1.02  3.36  0.73  0.95  3.71   0.90</span>
<span class="go">-----------------------------------</span>
<span class="go">=========================</span>
<span class="go">     edge  parent  length</span>
<span class="go">-------------------------</span>
<span class="go">    Human  edge.0    0.09</span>
<span class="go">HowlerMon  edge.0    0.12</span>
<span class="go">   edge.0  edge.1    0.12</span>
<span class="go">    Mouse  edge.1    0.84</span>
<span class="go">   edge.1    root    0.06</span>
<span class="go">NineBande    root    0.28</span>
<span class="go"> DogFaced    root    0.34</span>
<span class="go">-------------------------</span>
<span class="go">=============</span>
<span class="go">motif  mprobs</span>
<span class="go">-------------</span>
<span class="go">    T    0.23</span>
<span class="go">    C    0.19</span>
<span class="go">    A    0.37</span>
<span class="go">    G    0.21</span>
<span class="go">-------------</span>
</pre></div>
</div>
<p>We&#8217;ll get the lnL and number of free parameters for later use.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">null_lnL</span> <span class="o">=</span> <span class="n">lf</span><span class="o">.</span><span class="n">getLogLikelihood</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">null_nfp</span> <span class="o">=</span> <span class="n">lf</span><span class="o">.</span><span class="n">getNumFreeParams</span><span class="p">()</span>
</pre></div>
</div>
<p>Specify each edge has it&#8217;s own omega by just modifying the existing <tt class="docutils literal"><span class="pre">lf</span></tt>. This means the new function will start with the above values.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&quot;omega&quot;</span><span class="p">,</span> <span class="n">is_independent</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<p>Optimise the likelihood function, this time just using the local optimiser.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<p>View the resulting maximum-likelihood parameter values.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span>
<span class="go">Likelihood Function Table</span>
<span class="go">============================</span>
<span class="go"> A/C   A/G   A/T   C/G   C/T</span>
<span class="go">----------------------------</span>
<span class="go">1.03  3.38  0.73  0.95  3.72</span>
<span class="go">----------------------------</span>
<span class="go">================================</span>
<span class="go">     edge  parent  length  omega</span>
<span class="go">--------------------------------</span>
<span class="go">    Human  edge.0    0.09   0.59</span>
<span class="go">HowlerMon  edge.0    0.12   0.96</span>
<span class="go">   edge.0  edge.1    0.11   1.13</span>
<span class="go">    Mouse  edge.1    0.83   0.92</span>
<span class="go">   edge.1    root    0.06   0.39</span>
<span class="go">NineBande    root    0.28   1.28</span>
<span class="go"> DogFaced    root    0.34   0.84</span>
<span class="go">--------------------------------</span>
<span class="go">=============</span>
<span class="go">motif  mprobs</span>
<span class="go">-------------</span>
<span class="go">    T    0.23</span>
<span class="go">    C    0.19</span>
<span class="go">    A    0.37</span>
<span class="go">    G    0.21</span>
<span class="go">-------------</span>
</pre></div>
</div>
<p>Get out an annotated tree, it looks just like a tree, but has the maximum-likelihood parameter estimates attached to each tree edge. This object can be used for plotting, or to provide starting estimates to a related model.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">at</span> <span class="o">=</span> <span class="n">lf</span><span class="o">.</span><span class="n">getAnnotatedTree</span><span class="p">()</span>
</pre></div>
</div>
<p>The lnL&#8217;s from the two models are now used to calculate the likelihood ratio statistic (<tt class="docutils literal"><span class="pre">LR</span></tt>) it&#8217;s degrees-of-freedom (<tt class="docutils literal"><span class="pre">df</span></tt>) and the probability (<tt class="docutils literal"><span class="pre">P</span></tt>) of observing the LR.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">LR</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="p">(</span><span class="n">lf</span><span class="o">.</span><span class="n">getLogLikelihood</span><span class="p">()</span> <span class="o">-</span> <span class="n">null_lnL</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">lf</span><span class="o">.</span><span class="n">getNumFreeParams</span><span class="p">()</span> <span class="o">-</span> <span class="n">null_nfp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">P</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">chisqprob</span><span class="p">(</span><span class="n">LR</span><span class="p">,</span> <span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
<p>Print this and look up a chi-sq with number of edges - 1 degrees of freedom.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;Likelihood ratio statistic = &quot;</span><span class="p">,</span> <span class="n">LR</span>
<span class="go">Likelihood ratio statistic =  8...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;degrees-of-freedom = &quot;</span><span class="p">,</span> <span class="n">df</span>
<span class="go">degrees-of-freedom =  6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;probability = &quot;</span><span class="p">,</span> <span class="n">P</span>
<span class="go">probability =  0.2...</span>
</pre></div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div class="news">
<table id="feed"><tr><td><h3><a href="http://pycogent.wordpress.com/">PyCogent News and Announcements</a></h3></td>
</tr></table></div>

    


  <h4>Previous topic</h4>
  <p class="topless"><a href="relative_rate.html"
                        title="previous chapter">Performing a relative rate test</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="scope_model_params_on_trees.html"
                        title="next chapter">Allowing substitution model parameters to differ between branches</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/examples/neutral_test.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="scope_model_params_on_trees.html" title="Allowing substitution model parameters to differ between branches"
             >next</a> |</li>
        <li class="right" >
          <a href="relative_rate.html" title="Performing a relative rate test"
             >previous</a> |</li>
        <li><a href="../index.html">PyCogent 1.5.3 documentation</a> &raquo;</li>
          <li><a href="index.html" >Cogent Usage Examples</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2009, PyCogent Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>