<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Phylogenetic reconstruction by least squares &mdash; PyCogent 1.9 documentation</title>
    
    <link rel="stylesheet" href="../_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.9',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PyCogent 1.9 documentation" href="../index.html" />
    <link rel="up" title="Cogent Usage Examples" href="index.html" />
    <link rel="next" title="Making a phylogenetic tree from a protein sequence alignment" href="maketree_from_proteinseqs.html" />
    <link rel="prev" title="Make a UPGMA cluster" href="calculate_UPGMA_cluster.html" />     
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRQZ4X-jPDmWcshBrz2j7-fJvuUABRTGWmdiw2G89JpgztGlFGG8hDxRAw"></script>\
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body role="document">
    <div class="header-wrapper" role="banner">
      <div class="header">
        <div class="headertitle"><a
          href="../index.html">PyCogent 1.9 documentation</a></div>
        <div class="rel" role="navigation" aria-label="related navigation">
          <a href="calculate_UPGMA_cluster.html" title="Make a UPGMA cluster"
             accesskey="P">previous</a> |
          <a href="maketree_from_proteinseqs.html" title="Making a phylogenetic tree from a protein sequence alignment"
             accesskey="N">next</a> |
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="phylogenetic-reconstruction-by-least-squares">
<h1>Phylogenetic reconstruction by least squares<a class="headerlink" href="#phylogenetic-reconstruction-by-least-squares" title="Permalink to this headline">¶</a></h1>
<p><em>Section author: Gavin Huttley</em></p>
<p>We will load some pre-computed pairwise distance data. To see how that data was computed see the <a class="reference internal" href="calculate_pairwise_distances.html#calculating-pairwise-distances"><span class="std std-ref">Calculate pairwise distances between sequences</span></a> example. That data is saved in a format called <code class="docutils literal"><span class="pre">pickle</span></code> which is native to python. As per usual, we import the basic components we need.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">cPickle</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.phylo</span> <span class="k">import</span> <span class="n">distance</span><span class="p">,</span> <span class="n">least_squares</span>
</pre></div>
</div>
<p>Now load the distance data.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">file</span><span class="p">(</span><span class="s1">&#39;dists_for_phylo.pickle&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dists</span> <span class="o">=</span> <span class="n">cPickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>If there are extremely small distances, they can cause an error in the least squares calculation. Since such estimates are between extremely closely related sequences we could simply drop all distances for one of the sequences. We won&#8217;t do that here, we&#8217;ll leave that as exercise.</p>
<p>We make the ls calculator.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ls</span> <span class="o">=</span> <span class="n">least_squares</span><span class="o">.</span><span class="n">WLS</span><span class="p">(</span><span class="n">dists</span><span class="p">)</span>
</pre></div>
</div>
<p>We will search tree space for the collection of best trees using the advanced stepwise addition algorithm (hereafter <em>asaa</em>).</p>
<div class="section" id="look-for-the-single-best-tree">
<h2>Look for the single best tree<a class="headerlink" href="#look-for-the-single-best-tree" title="Permalink to this headline">¶</a></h2>
<p>In this use case we are after just 1 tree. We specify up to what taxa size all possible trees for the sample will be computed. Here we are specifying <code class="docutils literal"><span class="pre">a=5</span></code>. This means 5 sequences will be picked randomly and all possible trees relating them will be evaluated. <code class="docutils literal"><span class="pre">k=1</span></code> means only the best tree will be kept at the end of each such round of evaluation. For every remaining sequence it is grafted onto every possible branch of this tree. The best <code class="docutils literal"><span class="pre">k</span></code> results are then taken to the next round, when another sequence is randomly selected for addition. This proceeds until all sequences have been added. The result with following arguments is a single wls score and a single <code class="docutils literal"><span class="pre">Tree</span></code> which can be saved etc ..</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">score</span><span class="p">,</span> <span class="n">tree</span> <span class="o">=</span> <span class="n">ls</span><span class="o">.</span><span class="n">trex</span><span class="p">(</span><span class="n">a</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="n">score</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="n">e</span><span class="o">-</span><span class="mi">4</span>
</pre></div>
</div>
<p>We won&#8217;t display this tree, because we are doing more below.</p>
</div>
<div class="section" id="a-more-rigourous-tree-space-search">
<h2>A more rigourous tree space search<a class="headerlink" href="#a-more-rigourous-tree-space-search" title="Permalink to this headline">¶</a></h2>
<p>We change the asaa settings, so we keep more trees and then look at the distribution of the statistics for the last collection of trees. We could also change <code class="docutils literal"><span class="pre">a</span></code> to be larger, but in the current case we just adjust <code class="docutils literal"><span class="pre">k</span></code>. We also set the argument <code class="docutils literal"><span class="pre">return_all</span> <span class="pre">=</span> <span class="pre">True</span></code>, the effect of which is to return the complete set of saved trees. These, and their support statistic, can then be inspected.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">trees</span> <span class="o">=</span> <span class="n">ls</span><span class="o">.</span><span class="n">trex</span><span class="p">(</span><span class="n">a</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">return_all</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>Remember the sum-of-squares statistic will be smaller for &#8216;good&#8217; trees. The order of the trees returned is from good to bad. The number of returned <code class="docutils literal"><span class="pre">trees</span></code> is the same as the number requested to be retained at each step.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">trees</span><span class="p">)</span>
<span class="go">5</span>
</pre></div>
</div>
<p>Lets inspect the resulting statistics. First, the object <code class="docutils literal"><span class="pre">trees</span></code> is a list of <code class="docutils literal"><span class="pre">(wls,</span> <span class="pre">Tree)</span></code> tuples. We will therefore loop over the list to generate a separate list of just the wls statistics. The following syntax is called a list comprehension - basically just a very succinct <code class="docutils literal"><span class="pre">for</span></code> loop.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">wls_stats</span> <span class="o">=</span> <span class="p">[</span><span class="n">tree</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">trees</span><span class="p">]</span>
</pre></div>
</div>
<p>The <code class="docutils literal"><span class="pre">wls_stats</span></code> is a list which, if printed, looks like</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="p">[</span><span class="mf">1.3308768548934439e-05</span><span class="p">,</span> <span class="mf">0.0015588630350439783</span><span class="p">,</span> <span class="o">...</span>
</pre></div>
</div>
<p>From this you&#8217;ll see that the first 5 results are very similar to each other and would probably reasonably be considered equivalently supported topologies. I&#8217;ll just print the first two of the these trees after balancing them (in order to make their representations as equal as possible).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">t1</span> <span class="o">=</span> <span class="n">trees</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">balanced</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t2</span> <span class="o">=</span> <span class="n">trees</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">balanced</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">t1</span><span class="o">.</span><span class="n">asciiArt</span><span class="p">()</span>
<span class="go">                    /-Human</span>
<span class="go">          /edge.0--|</span>
<span class="go">         |          \-HowlerMon</span>
<span class="go">         |</span>
<span class="go">-root----|--Mouse</span>
<span class="go">         |</span>
<span class="go">         |          /-NineBande</span>
<span class="go">          \edge.1--|</span>
<span class="go">                    \-DogFaced</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">t2</span><span class="o">.</span><span class="n">asciiArt</span><span class="p">()</span>
<span class="go">          /-DogFaced</span>
<span class="go">         |</span>
<span class="go">         |          /-Human</span>
<span class="go">-root----|-edge.0--|</span>
<span class="go">         |          \-HowlerMon</span>
<span class="go">         |</span>
<span class="go">         |          /-NineBande</span>
<span class="go">          \edge.1--|</span>
<span class="go">                    \-Mouse</span>
</pre></div>
</div>
<p>You can see the difference involves the Jackrabbit, TreeShrew, Gorilla, Rat clade.</p>
</div>
<div class="section" id="assessing-the-fit-for-a-pre-specified-tree-topology">
<h2>Assessing the fit for a pre-specified tree topology<a class="headerlink" href="#assessing-the-fit-for-a-pre-specified-tree-topology" title="Permalink to this headline">¶</a></h2>
<p>In some instances we may have a tree from the literature or elsewhere whose fit to the data we seek to evaluate. In this case I&#8217;m going load a tree as follows.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="k">import</span> <span class="n">LoadTree</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query_tree</span> <span class="o">=</span> <span class="n">LoadTree</span><span class="p">(</span>
<span class="gp">... </span><span class="n">treestring</span><span class="o">=</span><span class="s2">&quot;((Human:.2,DogFaced:.2):.3,(NineBande:.1, Mouse:.5):.2,HowlerMon:.1)&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We now just use the <code class="docutils literal"><span class="pre">ls</span></code> object created above. The following evaluates the query using it&#8217;s associated branch lengths, returning only the wls statistic.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ls</span><span class="o">.</span><span class="n">evaluateTree</span><span class="p">(</span><span class="n">query_tree</span><span class="p">)</span>
<span class="go">2.8...</span>
</pre></div>
</div>
<p>We can also evaluate just the tree&#8217;s topology, returning both the wls statistic and the tree with best fit branch lengths.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">wls</span><span class="p">,</span> <span class="n">t</span> <span class="o">=</span> <span class="n">ls</span><span class="o">.</span><span class="n">evaluateTopology</span><span class="p">(</span><span class="n">query_tree</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="s2">&quot;</span><span class="si">%.4f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">wls</span> <span class="o">==</span> <span class="s1">&#39;0.0084&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="using-maximum-likelihood-for-measuring-tree-fit">
<h2>Using maximum likelihood for measuring tree fit<a class="headerlink" href="#using-maximum-likelihood-for-measuring-tree-fit" title="Permalink to this headline">¶</a></h2>
<p>This is a much slower algorithm and the interface largely mirrors that for the above. The difference is you import <code class="docutils literal"><span class="pre">maximum_likelihood</span></code> instead of <code class="docutils literal"><span class="pre">least_squares</span></code>, and use the <code class="docutils literal"><span class="pre">ML</span></code> instead of <code class="docutils literal"><span class="pre">WLS</span></code> classes. The <code class="docutils literal"><span class="pre">ML</span></code> class requires a substitution model (like a HKY85 for DNA or JTT92 for protein), and an alignment. It also optionally takes a distance matrix, such as that used here, computed for the same sequences. These distances are then used to obtain estimates of branch lengths by the WLS method for each evaluated tree topology which are then used as starting values for the likelihood optimisation.</p>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
<div class="news">
<table id="feed"><tr><td><h3><a href="http://pycogent.wordpress.com/">PyCogent News and Announcements</a></h3></td>
</tr></table></div>

    
          <h3>Table Of Contents</h3>
          <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Quick installation using pip</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html">The Readme</a></li>
<li class="toctree-l1"><a class="reference internal" href="../coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../data_file_links.html">The data files used in the documentation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Cogent Usage Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cookbook/index.html">PyCogent Cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_notes.html">For Developers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripting_guidelines.html">Scripting guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">Licenses and disclaimer</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ChangeLog.html">Changelog</a></li>
</ul>



          <div role="search">
            <h3 style="margin-top: 1.5em;">Search</h3>
            <form class="search" action="../search.html" method="get">
                <input type="text" name="q" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
            </form>
          </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <div role="navigation" aria-label="related navigaton">
            <a href="calculate_UPGMA_cluster.html" title="Make a UPGMA cluster"
              >previous</a> |
            <a href="maketree_from_proteinseqs.html" title="Making a phylogenetic tree from a protein sequence alignment"
              >next</a> |
            <a href="../genindex.html" title="General Index"
              >index</a>
          </div>
          <div role="note" aria-label="source link">
              <br/>
              <a href="../_sources/examples/phylo_by_ls.txt"
                rel="nofollow">Show Source</a>
          </div>
        </div>

        <div class="right">
          
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, PyCogent Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>