<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Reusing results to speed up optimisation &mdash; PyCogent 1.9 documentation</title>
    
    <link rel="stylesheet" href="../_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.9',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PyCogent 1.9 documentation" href="../index.html" />
    <link rel="up" title="Cogent Usage Examples" href="index.html" />
    <link rel="next" title="Specifying and using an unrestricted nucleotide substitution model" href="unrestricted_nucleotide.html" />
    <link rel="prev" title="Likelihood analysis of multiple loci" href="testing_multi_loci.html" />     
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRQZ4X-jPDmWcshBrz2j7-fJvuUABRTGWmdiw2G89JpgztGlFGG8hDxRAw"></script>\
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body role="document">
    <div class="header-wrapper" role="banner">
      <div class="header">
        <div class="headertitle"><a
          href="../index.html">PyCogent 1.9 documentation</a></div>
        <div class="rel" role="navigation" aria-label="related navigation">
          <a href="testing_multi_loci.html" title="Likelihood analysis of multiple loci"
             accesskey="P">previous</a> |
          <a href="unrestricted_nucleotide.html" title="Specifying and using an unrestricted nucleotide substitution model"
             accesskey="N">next</a> |
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="reusing-results-to-speed-up-optimisation">
<h1>Reusing results to speed up optimisation<a class="headerlink" href="#reusing-results-to-speed-up-optimisation" title="Permalink to this headline">Â¶</a></h1>
<p><em>Section author: Gavin Huttley</em></p>
<p>An example of how to use the maximum-likelihood parameter estimates from one model as starting values for another model. In this file we do something silly, by saving a result and then reloading it. This is silly because the analyses are run consecutively. A better approach when running consecutively is to simply use the annotated tree directly.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="k">import</span> <span class="n">LoadSeqs</span><span class="p">,</span> <span class="n">LoadTree</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.evolve.models</span> <span class="k">import</span> <span class="n">MG94HKY</span>
</pre></div>
</div>
<p>We&#8217;ll create a simple model, optimise it and save it for later reuse</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">=</span> <span class="n">LoadTree</span><span class="p">(</span><span class="s2">&quot;data/test.tree&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span> <span class="o">=</span> <span class="n">MG94HKY</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">makeLikelihoodFunction</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">digits</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">space</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setAlignment</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">lf</span>
<span class="go">Likelihood Function Table</span>
<span class="go">============</span>
<span class="go">kappa  omega</span>
<span class="go">------------</span>
<span class="go"> 3.85   0.90</span>
<span class="go">------------</span>
<span class="go">=========================</span>
<span class="go">     edge  parent  length</span>
<span class="go">-------------------------</span>
<span class="go">    Human  edge.0    0.09</span>
<span class="go">HowlerMon  edge.0    0.12</span>
<span class="go">   edge.0  edge.1    0.12</span>
<span class="go">    Mouse  edge.1    0.84</span>
<span class="go">   edge.1    root    0.06</span>
<span class="go">NineBande    root    0.28</span>
<span class="go"> DogFaced    root    0.34</span>
<span class="go">-------------------------</span>
<span class="go">=============</span>
<span class="go">motif  mprobs</span>
<span class="go">-------------</span>
<span class="go">    T    0.23</span>
<span class="go">    C    0.19</span>
<span class="go">    A    0.37</span>
<span class="go">    G    0.21</span>
<span class="go">-------------</span>
</pre></div>
</div>
<p>The essential object for reuse is an annotated tree these capture the parameter estimates from the above optimisation we can either use this directly in the same run, or we can save the tree to file in <code class="docutils literal"><span class="pre">xml</span></code> format and reload the tree at a later time for use. In this example I&#8217;ll illustrate the latter scenario.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">at</span><span class="o">=</span><span class="n">lf</span><span class="o">.</span><span class="n">getAnnotatedTree</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">at</span><span class="o">.</span><span class="n">writeToFile</span><span class="p">(</span><span class="s1">&#39;tree.xml&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>We load the tree as per usual</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">nt</span> <span class="o">=</span> <span class="n">LoadTree</span><span class="p">(</span><span class="s1">&#39;tree.xml&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Now create a more parameter rich model, in this case by allowing the <code class="docutils literal"><span class="pre">Human</span></code> edge to have a different value of <code class="docutils literal"><span class="pre">omega</span></code>. By providing the annotated tree, the parameter estimates from the above run will be used as starting values for the new model.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">new_lf</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">makeLikelihoodFunction</span><span class="p">(</span><span class="n">nt</span><span class="p">,</span> <span class="n">digits</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">space</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_lf</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s1">&#39;omega&#39;</span><span class="p">,</span> <span class="n">edge</span><span class="o">=</span><span class="s1">&#39;Human&#39;</span><span class="p">,</span>
<span class="gp">... </span>                    <span class="n">is_independent</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_lf</span><span class="o">.</span><span class="n">setAlignment</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">new_lf</span>
<span class="go">Likelihood Function Table</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 3.85</span>
<span class="go">-----</span>
<span class="go">================================</span>
<span class="go">     edge  parent  length  omega</span>
<span class="go">--------------------------------</span>
<span class="go">    Human  edge.0    0.09   0.59</span>
<span class="go">HowlerMon  edge.0    0.12   0.92</span>
<span class="go">   edge.0  edge.1    0.12   0.92</span>
<span class="go">    Mouse  edge.1    0.84   0.92</span>
<span class="go">   edge.1    root    0.06   0.92</span>
<span class="go">NineBande    root    0.28   0.92</span>
<span class="go"> DogFaced    root    0.34   0.92</span>
<span class="go">--------------------------------</span>
<span class="go">=============</span>
<span class="go">motif  mprobs</span>
<span class="go">-------------</span>
<span class="go">    T    0.23</span>
<span class="go">    C    0.19</span>
<span class="go">    A    0.37</span>
<span class="go">    G    0.21</span>
<span class="go">-------------</span>
</pre></div>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
<div class="news">
<table id="feed"><tr><td><h3><a href="http://pycogent.wordpress.com/">PyCogent News and Announcements</a></h3></td>
</tr></table></div>

    
          <h3>Table Of Contents</h3>
          <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Quick installation using pip</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html">The Readme</a></li>
<li class="toctree-l1"><a class="reference internal" href="../coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../data_file_links.html">The data files used in the documentation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Cogent Usage Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cookbook/index.html">PyCogent Cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_notes.html">For Developers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripting_guidelines.html">Scripting guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">Licenses and disclaimer</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ChangeLog.html">Changelog</a></li>
</ul>



          <div role="search">
            <h3 style="margin-top: 1.5em;">Search</h3>
            <form class="search" action="../search.html" method="get">
                <input type="text" name="q" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
            </form>
          </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <div role="navigation" aria-label="related navigaton">
            <a href="testing_multi_loci.html" title="Likelihood analysis of multiple loci"
              >previous</a> |
            <a href="unrestricted_nucleotide.html" title="Specifying and using an unrestricted nucleotide substitution model"
              >next</a> |
            <a href="../genindex.html" title="General Index"
              >index</a>
          </div>
          <div role="note" aria-label="source link">
              <br/>
              <a href="../_sources/examples/reuse_results.txt"
                rel="nofollow">Show Source</a>
          </div>
        </div>

        <div class="right">
          
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, PyCogent Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>