

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Allowing substitution model parameters to differ between branches &mdash; PyCogent 1.5.3 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.5.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PyCogent 1.5.3 documentation" href="../index.html" />
    <link rel="up" title="Cogent Usage Examples" href="index.html" />
    <link rel="next" title="Using codon models" href="codon_models.html" />
    <link rel="prev" title="A test of the neutral theory" href="neutral_test.html" />     
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRQZ4X-jPDmWcshBrz2j7-fJvuUABRTGWmdiw2G89JpgztGlFGG8hDxRAw"></script>\
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="codon_models.html" title="Using codon models"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="neutral_test.html" title="A test of the neutral theory"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">PyCogent 1.5.3 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Cogent Usage Examples</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="allowing-substitution-model-parameters-to-differ-between-branches">
<span id="scope-params-on-trees"></span><h1>Allowing substitution model parameters to differ between branches<a class="headerlink" href="#allowing-substitution-model-parameters-to-differ-between-branches" title="Permalink to this headline">¶</a></h1>
<p><em>Section author: Gavin Huttley</em></p>
<p>A common task concerns assessing how substitution model exchangeability parameters differ between evolutionary lineages. This is most commonly of interest for the case of testing for natural selection. Here I&#8217;ll demonstrate the different ways of scoping parameters across trees for the codon model case and how these can be used for evolutionary modelling.</p>
<p>We start with the standard imports, plus using a canned codon substitution model and then load the sample data set.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span><span class="p">,</span> <span class="n">LoadTree</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.evolve.models</span> <span class="kn">import</span> <span class="n">MG94HKY</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="s">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">LoadTree</span><span class="p">(</span><span class="s">&quot;data/test.tree&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We construct the substitution model and likelihood function and set the alignment.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span> <span class="o">=</span> <span class="n">MG94HKY</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">makeLikelihoodFunction</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">digits</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">space</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setAlignment</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
</pre></div>
</div>
<p>At this point we have a likelihood function with two exchangeability parameters from the substitution model (<tt class="docutils literal"><span class="pre">kappa</span></tt> the transition/transversion ratio; <tt class="docutils literal"><span class="pre">omega</span></tt> the nonsynonymous/synonymous ratio) plus branch lengths for all tree edges. To facilitate subsequent discussion I now display the tree</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">tree</span><span class="o">.</span><span class="n">asciiArt</span><span class="p">()</span>
<span class="go">                              /-Human</span>
<span class="go">                    /edge.0--|</span>
<span class="go">          /edge.1--|          \-HowlerMon</span>
<span class="go">         |         |</span>
<span class="go">         |          \-Mouse</span>
<span class="go">-root----|</span>
<span class="go">         |--NineBande</span>
<span class="go">         |</span>
<span class="go">          \-DogFaced</span>
</pre></div>
</div>
<p>In order to scope a parameter on a tree (meaning specifying a subset of edges for which the parameter is to be treated differently to the remainder of the tree) requires uniquely identifying the edges. We do this using the following arguments to the likelihood function <tt class="docutils literal"><span class="pre">setParamRule</span></tt> method:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">tip_names</span></tt>: the name of two tips</li>
<li><tt class="docutils literal"><span class="pre">outgroup_name</span></tt>: the name of a tip that is not part of the clade of interest</li>
<li><tt class="docutils literal"><span class="pre">is_clade</span></tt>: if <tt class="docutils literal"><span class="pre">True</span></tt>, all lineages descended from the tree node identified by the <tt class="docutils literal"><span class="pre">tip_names</span></tt> and <tt class="docutils literal"><span class="pre">outgroup_name</span></tt> argument are affected by the other arguments. If <tt class="docutils literal"><span class="pre">False</span></tt>, then the <tt class="docutils literal"><span class="pre">is_stem</span></tt> argument must apply.</li>
<li><tt class="docutils literal"><span class="pre">is_stem</span></tt>: Whether the edge leading to the node is included.</li>
</ul>
<p>The next concepts include exactly what can be scoped and how. In the case of testing for distinctive periods of natural selection it is common to specify distinct values for <tt class="docutils literal"><span class="pre">omega</span></tt> for an edge. I&#8217;ll first illustrate some possible uses for the arguments above by setting <tt class="docutils literal"><span class="pre">omega</span></tt> to be distinctive for specific edges. I will set a value for omega so that printing the likelihood function illustrates what edges have been effected, but I won&#8217;t actually do any model fitting.</p>
<div class="section" id="specifying-a-clade">
<h2>Specifying a clade<a class="headerlink" href="#specifying-a-clade" title="Permalink to this headline">¶</a></h2>
<p>I&#8217;m going to cause <tt class="docutils literal"><span class="pre">omega</span></tt> to attain a different value for all branches aside from the primate clade and stem (<tt class="docutils literal"><span class="pre">HowlerMon</span></tt>, <tt class="docutils literal"><span class="pre">Human</span></tt>, <tt class="docutils literal"><span class="pre">edge.0</span></tt>).</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;DogFaced&#39;</span><span class="p">,</span> <span class="s">&#39;Mouse&#39;</span><span class="p">],</span>
<span class="gp">... </span>             <span class="n">outgroup_name</span><span class="o">=</span><span class="s">&#39;Human&#39;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">is_clade</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span>
<span class="go">Likelihood Function Table</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 1.00</span>
<span class="go">-----</span>
<span class="go">===================================</span>
<span class="go">     edge   parent   length   omega</span>
<span class="go">-----------------------------------</span>
<span class="go">    Human   edge.0     0.03    1.00</span>
<span class="go">HowlerMon   edge.0     0.04    1.00</span>
<span class="go">   edge.0   edge.1     0.04    1.00</span>
<span class="go">    Mouse   edge.1     0.28    2.00</span>
<span class="go">   edge.1     root     0.02    2.00</span>
<span class="go">NineBande     root     0.09    2.00</span>
<span class="go"> DogFaced     root     0.11    2.00</span>
<span class="go">-----------------------------------</span>
<span class="go">==============</span>
<span class="go">motif   mprobs</span>
<span class="go">--------------</span>
<span class="go">    T     0.23</span>
<span class="go">    C     0.19</span>
<span class="go">    A     0.37</span>
<span class="go">    G     0.21</span>
<span class="go">--------------</span>
</pre></div>
</div>
<p>As you can see <tt class="docutils literal"><span class="pre">omega</span></tt> for the primate edges I listed above have the default parameter value (1.0), while the others have what I&#8217;ve assigned. In fact, you could omit the <tt class="docutils literal"><span class="pre">is_clade</span></tt> argument as this is the default, but I think for readability of scripts it&#8217;s best to be explicit.</p>
</div>
<div class="section" id="specifying-a-stem">
<h2>Specifying a stem<a class="headerlink" href="#specifying-a-stem" title="Permalink to this headline">¶</a></h2>
<p>This time I&#8217;ll specify the stem leading to the primates as the edge of interest.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">I need to reset the <tt class="docutils literal"><span class="pre">lf</span></tt> so all edges have the default value again. I&#8217;ll show this only for this example, but rest assured I&#8217;m doing it for all others too.</p>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;omega&#39;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;Human&#39;</span><span class="p">,</span> <span class="s">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">is_stem</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">is_clade</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span>
<span class="go">Likelihood Function Table</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 1.00</span>
<span class="go">-----</span>
<span class="go">===================================</span>
<span class="go">     edge   parent   length   omega</span>
<span class="go">-----------------------------------</span>
<span class="go">    Human   edge.0     0.03    1.00</span>
<span class="go">HowlerMon   edge.0     0.04    1.00</span>
<span class="go">   edge.0   edge.1     0.04    2.00</span>
<span class="go">    Mouse   edge.1     0.28    1.00</span>
<span class="go">   edge.1     root     0.02    1.00</span>
<span class="go">NineBande     root     0.09    1.00</span>
<span class="go"> DogFaced     root     0.11    1.00</span>
<span class="go">-----------------------------------...</span>
</pre></div>
</div>
</div>
<div class="section" id="specifying-clade-and-stem">
<h2>Specifying clade and stem<a class="headerlink" href="#specifying-clade-and-stem" title="Permalink to this headline">¶</a></h2>
<p>I&#8217;ll specify that both the primates and their stem are to be considered.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;Human&#39;</span><span class="p">,</span> <span class="s">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">is_stem</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">is_clade</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span>
<span class="go">Likelihood Function Table</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 1.00</span>
<span class="go">-----</span>
<span class="go">===================================</span>
<span class="go">     edge   parent   length   omega</span>
<span class="go">-----------------------------------</span>
<span class="go">    Human   edge.0     0.03    2.00</span>
<span class="go">HowlerMon   edge.0     0.04    2.00</span>
<span class="go">   edge.0   edge.1     0.04    2.00</span>
<span class="go">    Mouse   edge.1     0.28    1.00</span>
<span class="go">   edge.1     root     0.02    1.00</span>
<span class="go">NineBande     root     0.09    1.00</span>
<span class="go"> DogFaced     root     0.11    1.00</span>
<span class="go">-----------------------------------...</span>
</pre></div>
</div>
</div>
<div class="section" id="alternate-arguments-for-specifying-edges">
<h2>Alternate arguments for specifying edges<a class="headerlink" href="#alternate-arguments-for-specifying-edges" title="Permalink to this headline">¶</a></h2>
<p>The likelihood function <tt class="docutils literal"><span class="pre">setParamRule</span></tt> method also has the arguments of <tt class="docutils literal"><span class="pre">edge</span></tt> and <tt class="docutils literal"><span class="pre">edges</span></tt>. These allow specific naming of the tree edge(s) to be affected by a rule. In general, however, the <tt class="docutils literal"><span class="pre">tip_names</span></tt> + <tt class="docutils literal"><span class="pre">outgroup_name</span></tt> combo is more robust.</p>
</div>
<div class="section" id="applications-of-scoped-parameters">
<h2>Applications of scoped parameters<a class="headerlink" href="#applications-of-scoped-parameters" title="Permalink to this headline">¶</a></h2>
<p>The general use-cases for which a tree scope can be applied are:</p>
<ol class="arabic">
<li><p class="first">constraining all edges identified by a rule to have a specific value which is constant and not modifiable</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;Human&#39;</span><span class="p">,</span> <span class="s">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">is_clade</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">is_constant</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p class="first">all edges identified by a rule have the same but different value to the rest of the tree</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;Human&#39;</span><span class="p">,</span> <span class="s">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">is_clade</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p class="first">allowing all edges identified by a rule to have different values of the parameter with the remaining tree edges having the same value</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;Human&#39;</span><span class="p">,</span> <span class="s">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">is_clade</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">is_independent</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p class="first">allowing all edges to have a different value</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;omega&#39;</span><span class="p">,</span> <span class="n">is_independent</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
</li>
</ol>
<p>I&#8217;ll demonstrate these cases sequentially as they involve gradually increasing the degrees of freedom in the model. First we&#8217;ll constrain <tt class="docutils literal"><span class="pre">omega</span></tt> to equal 1 on the primate edges. I&#8217;ll then optimise the model.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">here I&#8217;m specifying a constant value for the parameter and so I <strong>must</strong> use the argument <tt class="docutils literal"><span class="pre">value</span></tt> to set it. This not to be confused with the argument <tt class="docutils literal"><span class="pre">init</span></tt> that is used for providing initial (starting) values for fitting.</p>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;Human&#39;</span><span class="p">,</span> <span class="s">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">is_clade</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">is_constant</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span>
<span class="go">Likelihood Function Table</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 3.87</span>
<span class="go">-----</span>
<span class="go">===================================</span>
<span class="go">     edge   parent   length   omega</span>
<span class="go">-----------------------------------</span>
<span class="go">    Human   edge.0     0.09    1.00</span>
<span class="go">HowlerMon   edge.0     0.12    1.00</span>
<span class="go">   edge.0   edge.1     0.12    0.92</span>
<span class="go">    Mouse   edge.1     0.84    0.92</span>
<span class="go">   edge.1     root     0.06    0.92</span>
<span class="go">NineBande     root     0.28    0.92</span>
<span class="go"> DogFaced     root     0.34    0.92</span>
<span class="go">-----------------------------------</span>
<span class="go">==============</span>
<span class="go">motif   mprobs</span>
<span class="go">--------------</span>
<span class="go">    T     0.23</span>
<span class="go">    C     0.19</span>
<span class="go">    A     0.37</span>
<span class="go">    G     0.21</span>
<span class="go">--------------</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span><span class="o">.</span><span class="n">getLogLikelihood</span><span class="p">()</span>
<span class="go">-8640.9...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span><span class="o">.</span><span class="n">getNumFreeParams</span><span class="p">()</span>
<span class="go">9</span>
</pre></div>
</div>
<p>I&#8217;ll now free up <tt class="docutils literal"><span class="pre">omega</span></tt> on the primate clade, but making it a single value shared by all primate lineages.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;Human&#39;</span><span class="p">,</span> <span class="s">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">is_clade</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">is_constant</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span>
<span class="go">Likelihood Function Table</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 3.85</span>
<span class="go">-----</span>
<span class="go">===================================</span>
<span class="go">     edge   parent   length   omega</span>
<span class="go">-----------------------------------</span>
<span class="go">    Human   edge.0     0.09    0.77</span>
<span class="go">HowlerMon   edge.0     0.12    0.77</span>
<span class="go">   edge.0   edge.1     0.12    0.92</span>
<span class="go">    Mouse   edge.1     0.84    0.92</span>
<span class="go">   edge.1     root     0.06    0.92</span>
<span class="go">NineBande     root     0.28    0.92</span>
<span class="go"> DogFaced     root     0.34    0.92</span>
<span class="go">-----------------------------------</span>
<span class="go">==============</span>
<span class="go">motif   mprobs</span>
<span class="go">--------------</span>
<span class="go">    T     0.23</span>
<span class="go">    C     0.19</span>
<span class="go">    A     0.37</span>
<span class="go">    G     0.21</span>
<span class="go">--------------</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span><span class="o">.</span><span class="n">getLogLikelihood</span><span class="p">()</span>
<span class="go">-8639.7...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span><span class="o">.</span><span class="n">getNumFreeParams</span><span class="p">()</span>
<span class="go">10</span>
</pre></div>
</div>
<p>Finally I&#8217;ll allow all primate edges to have different values of <tt class="docutils literal"><span class="pre">omega</span></tt>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;Human&#39;</span><span class="p">,</span> <span class="s">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">is_clade</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">is_independent</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span>
<span class="go">Likelihood Function Table</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 3.85</span>
<span class="go">-----</span>
<span class="go">===================================</span>
<span class="go">     edge   parent   length   omega</span>
<span class="go">-----------------------------------</span>
<span class="go">    Human   edge.0     0.09    0.59</span>
<span class="go">HowlerMon   edge.0     0.12    0.95</span>
<span class="go">   edge.0   edge.1     0.12    0.92</span>
<span class="go">    Mouse   edge.1     0.84    0.92</span>
<span class="go">   edge.1     root     0.06    0.92</span>
<span class="go">NineBande     root     0.28    0.92</span>
<span class="go"> DogFaced     root     0.34    0.92</span>
<span class="go">-----------------------------------</span>
<span class="go">==============</span>
<span class="go">motif   mprobs</span>
<span class="go">--------------</span>
<span class="go">    T     0.23</span>
<span class="go">    C     0.19</span>
<span class="go">    A     0.37</span>
<span class="go">    G     0.21</span>
<span class="go">--------------</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span><span class="o">.</span><span class="n">getLogLikelihood</span><span class="p">()</span>
<span class="go">-8638.9...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span><span class="o">.</span><span class="n">getNumFreeParams</span><span class="p">()</span>
<span class="go">11</span>
</pre></div>
</div>
<p>We now allow <tt class="docutils literal"><span class="pre">omega</span></tt> to be different on all edges.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">setParamRule</span><span class="p">(</span><span class="s">&#39;omega&#39;</span><span class="p">,</span> <span class="n">is_independent</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span>
<span class="go">Likelihood Function Table</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 3.85</span>
<span class="go">-----</span>
<span class="go">===================================</span>
<span class="go">     edge   parent   length   omega</span>
<span class="go">-----------------------------------</span>
<span class="go">    Human   edge.0     0.09    0.59</span>
<span class="go">HowlerMon   edge.0     0.12    0.95</span>
<span class="go">   edge.0   edge.1     0.12    1.13</span>
<span class="go">    Mouse   edge.1     0.84    0.92</span>
<span class="go">   edge.1     root     0.06    0.38</span>
<span class="go">NineBande     root     0.28    1.27</span>
<span class="go"> DogFaced     root     0.34    0.84</span>
<span class="go">-----------------------------------</span>
<span class="go">==============</span>
<span class="go">motif   mprobs</span>
<span class="go">--------------</span>
<span class="go">    T     0.23</span>
<span class="go">    C     0.19</span>
<span class="go">    A     0.37</span>
<span class="go">    G     0.21</span>
<span class="go">--------------</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span><span class="o">.</span><span class="n">getLogLikelihood</span><span class="p">()</span>
<span class="go">-8636.1...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">lf</span><span class="o">.</span><span class="n">getNumFreeParams</span><span class="p">()</span>
<span class="go">15</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div class="news">
<table id="feed"><tr><td><h3><a href="http://pycogent.wordpress.com/">PyCogent News and Announcements</a></h3></td>
</tr></table></div>

    
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Allowing substitution model parameters to differ between branches</a><ul>
<li><a class="reference internal" href="#specifying-a-clade">Specifying a clade</a></li>
<li><a class="reference internal" href="#specifying-a-stem">Specifying a stem</a></li>
<li><a class="reference internal" href="#specifying-clade-and-stem">Specifying clade and stem</a></li>
<li><a class="reference internal" href="#alternate-arguments-for-specifying-edges">Alternate arguments for specifying edges</a></li>
<li><a class="reference internal" href="#applications-of-scoped-parameters">Applications of scoped parameters</a></li>
</ul>
</li>
</ul>



  <h4>Previous topic</h4>
  <p class="topless"><a href="neutral_test.html"
                        title="previous chapter">A test of the neutral theory</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="codon_models.html"
                        title="next chapter">Using codon models</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/examples/scope_model_params_on_trees.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="codon_models.html" title="Using codon models"
             >next</a> |</li>
        <li class="right" >
          <a href="neutral_test.html" title="A test of the neutral theory"
             >previous</a> |</li>
        <li><a href="../index.html">PyCogent 1.5.3 documentation</a> &raquo;</li>
          <li><a href="index.html" >Cogent Usage Examples</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2009, PyCogent Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>