<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Seqsim Simple Alignment Simulation Example &mdash; PyCogent 1.9 documentation</title>
    
    <link rel="stylesheet" href="../_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.9',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PyCogent 1.9 documentation" href="../index.html" />
    <link rel="up" title="Cogent Usage Examples" href="index.html" />
    <link rel="next" title="Seqsim Alignment Simulation Example with Non-standard alphabet" href="seqsim_aln_sim_user_alphabet.html" />
    <link rel="prev" title="Evaluate process heterogeneity using a Hidden Markov Model" href="hmm_par_heterogeneity.html" />     
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRQZ4X-jPDmWcshBrz2j7-fJvuUABRTGWmdiw2G89JpgztGlFGG8hDxRAw"></script>\
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body role="document">
    <div class="header-wrapper" role="banner">
      <div class="header">
        <div class="headertitle"><a
          href="../index.html">PyCogent 1.9 documentation</a></div>
        <div class="rel" role="navigation" aria-label="related navigation">
          <a href="hmm_par_heterogeneity.html" title="Evaluate process heterogeneity using a Hidden Markov Model"
             accesskey="P">previous</a> |
          <a href="seqsim_aln_sim_user_alphabet.html" title="Seqsim Alignment Simulation Example with Non-standard alphabet"
             accesskey="N">next</a> |
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="seqsim-simple-alignment-simulation-example">
<h1>Seqsim Simple Alignment Simulation Example<a class="headerlink" href="#seqsim-simple-alignment-simulation-example" title="Permalink to this headline">Â¶</a></h1>
<p><em>Section author: Julia Goodrich</em></p>
<p>This is a very simple example of how to use the <code class="docutils literal"><span class="pre">seqsim</span></code> module to simulate an alignment for a tree starting with a random sequence and substitution rate matrix (q). The rate matrix gives the rate constant of going from one character in the sequence to another character in the sequence, the Q matrix determines the rate of change of the sequence.</p>
<p>First we will perform the necessary imports:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">Rates</span></code> is an object that stores the rate matrix data, it can also be used to generate a random rate matrix given an <code class="docutils literal"><span class="pre">Alphabet</span></code>.</li>
<li><code class="docutils literal"><span class="pre">DnaUsage</span></code> is a <code class="docutils literal"><span class="pre">Usage</span></code> object that stores the usage of each nucleotide.</li>
<li><code class="docutils literal"><span class="pre">DnaPairs</span></code> is an Alphabet it stores the DNA pairs (AA,AT,AC,...), it can be passed into the <code class="docutils literal"><span class="pre">Rates</span></code> object, defining the rate matrix pairs for DNA.</li>
<li><code class="docutils literal"><span class="pre">DNA</span></code> is a <code class="docutils literal"><span class="pre">MolType</span></code> object for DNA.</li>
<li><code class="docutils literal"><span class="pre">RangeNode</span></code> is the main <code class="docutils literal"><span class="pre">seqsim</span></code> Node object, it allows for the easy evolution of sequences.</li>
<li><code class="docutils literal"><span class="pre">DndParser</span></code> is a parser for a newick format tree.</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.seqsim.usage</span> <span class="k">import</span> <span class="n">Rates</span><span class="p">,</span> <span class="n">DnaUsage</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.core.usage</span> <span class="k">import</span> <span class="n">DnaPairs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.core.moltype</span> <span class="k">import</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.core.alignment</span> <span class="k">import</span> <span class="n">Alignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.seqsim.tree</span> <span class="k">import</span> <span class="n">RangeNode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.parse.tree</span> <span class="k">import</span> <span class="n">DndParser</span>
</pre></div>
</div>
<p>Now, lets specify a 4 taxon tree:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">=</span> <span class="n">DndParser</span><span class="p">(</span><span class="s1">&#39;(a:0.4,b:0.3,(c:0.15,d:0.2)edge.0:0.1);&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">constructor</span> <span class="o">=</span> <span class="n">RangeNode</span><span class="p">)</span>
</pre></div>
</div>
<p>To generate a random DNA sequence, we first specify nucleotide frequencies with the <code class="docutils literal"><span class="pre">DnaUsage</span></code> object. Then we create a random DNA sequence that is five bases long.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">=</span> <span class="n">DnaUsage</span><span class="p">({</span><span class="s1">&#39;A&#39;</span><span class="p">:</span><span class="mf">0.5</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">:</span><span class="mf">0.2</span><span class="p">,</span><span class="s1">&#39;C&#39;</span><span class="p">:</span><span class="mf">0.15</span><span class="p">,</span><span class="s1">&#39;G&#39;</span><span class="p">:</span><span class="mf">0.25</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">ModelSeq</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">randomIndices</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">q</span> <span class="o">=</span> <span class="n">Rates</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">DnaPairs</span><span class="p">)</span>
</pre></div>
</div>
<p>Set q at the base of the tree and propagate it to all nodes in the tree,</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">Q</span> <span class="o">=</span> <span class="n">q</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">propagateAttr</span><span class="p">(</span><span class="s1">&#39;Q&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Set a P matrix (probability matrix) from every Q matrix on each node, <img class="math" src="../_images/math/eb6ea99966bb182a4cba4f92dbf8e1e2641f27ed.png" alt="P(t) = e^(Qt)"/>,</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">assignP</span><span class="p">()</span>
</pre></div>
</div>
<p>Use <code class="docutils literal"><span class="pre">evolve</span></code> to evolve sequences for each tip, Note: must evolve sequence
data, not sequence object itself (for speed)</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">_data</span><span class="p">)</span>
</pre></div>
</div>
<p>Build alignment,</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">{}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">tips</span><span class="p">():</span>
<span class="gp">... </span>    <span class="n">seqs</span><span class="p">[</span><span class="n">n</span><span class="o">.</span><span class="n">Name</span><span class="p">]</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">ModelSeq</span><span class="p">(</span><span class="n">n</span><span class="o">.</span><span class="n">Sequence</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
</pre></div>
</div>
<p>The result is a Cogent <code class="docutils literal"><span class="pre">Alignment</span></code> object, which can be used the same way as any other alignment object.</p>
<p><code class="docutils literal"><span class="pre">evolveSeqs</span></code> can be used instead of evolve to evolve multiple sequences according to the same tree (can model either different genes, or different rate categories within a gene that you then combine, etc...),</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">numpy</span> <span class="k">import</span> <span class="n">concatenate</span>
</pre></div>
</div>
<p>First you need to use <code class="docutils literal"><span class="pre">assignPs</span></code> to assign the proper P matricies given rates:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">assignPs</span><span class="p">([</span><span class="o">.</span><span class="mi">5</span><span class="p">,</span> <span class="o">.</span><span class="mi">75</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
<p>There needs to be the same number of random sequences as there are rate catigories so we create a list of 3 random sequences,</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">[</span><span class="n">DNA</span><span class="o">.</span><span class="n">ModelSeq</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">randomIndices</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span><span class="o">.</span><span class="n">_data</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">3</span><span class="p">)]</span>
</pre></div>
</div>
<p>Then use <code class="docutils literal"><span class="pre">evolveSeqs</span></code> to evolve a sequence for every tip with every rate.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">evolveSeqs</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
</pre></div>
</div>
<p>Now to concatenate the sequences,</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">{}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">tips</span><span class="p">():</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">n</span><span class="o">.</span><span class="n">Sequences</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">seqs</span><span class="p">[</span><span class="n">n</span><span class="o">.</span><span class="n">Name</span><span class="p">]</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">ModelSeq</span><span class="p">(</span><span class="n">concatenate</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">n</span><span class="o">.</span><span class="n">Sequences</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
</pre></div>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
<div class="news">
<table id="feed"><tr><td><h3><a href="http://pycogent.wordpress.com/">PyCogent News and Announcements</a></h3></td>
</tr></table></div>

    
          <h3>Table Of Contents</h3>
          <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Quick installation using pip</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html">The Readme</a></li>
<li class="toctree-l1"><a class="reference internal" href="../coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../data_file_links.html">The data files used in the documentation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Cogent Usage Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cookbook/index.html">PyCogent Cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_notes.html">For Developers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripting_guidelines.html">Scripting guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">Licenses and disclaimer</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ChangeLog.html">Changelog</a></li>
</ul>



          <div role="search">
            <h3 style="margin-top: 1.5em;">Search</h3>
            <form class="search" action="../search.html" method="get">
                <input type="text" name="q" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
            </form>
          </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <div role="navigation" aria-label="related navigaton">
            <a href="hmm_par_heterogeneity.html" title="Evaluate process heterogeneity using a Hidden Markov Model"
              >previous</a> |
            <a href="seqsim_aln_sim_user_alphabet.html" title="Seqsim Alignment Simulation Example with Non-standard alphabet"
              >next</a> |
            <a href="../genindex.html" title="General Index"
              >index</a>
          </div>
          <div role="note" aria-label="source link">
              <br/>
              <a href="../_sources/examples/seqsim_alignment_simulation.txt"
                rel="nofollow">Show Source</a>
          </div>
        </div>

        <div class="right">
          
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, PyCogent Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>