

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Building phylogenies &mdash; PyCogent 1.5.3 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.5.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PyCogent 1.5.3 documentation" href="../index.html" />
    <link rel="up" title="PyCogent Cookbook" href="index.html" />
    <link rel="next" title="Evolutionary analysis using likelihood" href="using_likelihood_to_perform_evolutionary_analyses.html" />
    <link rel="prev" title="Building a tree of life" href="building_a_tree_of_life.html" />     
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRQZ4X-jPDmWcshBrz2j7-fJvuUABRTGWmdiw2G89JpgztGlFGG8hDxRAw"></script>\
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="using_likelihood_to_perform_evolutionary_analyses.html" title="Evolutionary analysis using likelihood"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="building_a_tree_of_life.html" title="Building a tree of life"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">PyCogent 1.5.3 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">PyCogent Cookbook</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="building-phylogenies">
<h1>Building phylogenies<a class="headerlink" href="#building-phylogenies" title="Permalink to this headline">¶</a></h1>
<div class="section" id="built-in-phylogenetic-reconstruction">
<h2>Built-in Phylogenetic reconstruction<a class="headerlink" href="#built-in-phylogenetic-reconstruction" title="Permalink to this headline">¶</a></h2>
<div class="section" id="by-distance-method">
<h3>By distance method<a class="headerlink" href="#by-distance-method" title="Permalink to this headline">¶</a></h3>
<p>Given an alignment, a phylogenetic tree can be generated based on the pair-wise distance matrix computed from the alignment.</p>
<div class="section" id="fast-pairwise-distance-estimation">
<h4>Fast pairwise distance estimation<a class="headerlink" href="#fast-pairwise-distance-estimation" title="Permalink to this headline">¶</a></h4>
<p>For a limited number of evolutionary models a fast implementation is available. Here we use the Tamura and Nei 1993 model.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.evolve.pairwise_distance</span> <span class="kn">import</span> <span class="n">TN93Pair</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="s">&#39;data/primate_brca1.fasta&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dist_calc</span> <span class="o">=</span> <span class="n">TN93Pair</span><span class="p">(</span><span class="n">DNA</span><span class="p">,</span> <span class="n">alignment</span><span class="o">=</span><span class="n">aln</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dist_calc</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>We can obtain the distances as a <tt class="docutils literal"><span class="pre">dict</span></tt> for direct usage in phylogenetic reconstruction</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">dists</span> <span class="o">=</span> <span class="n">dist_calc</span><span class="o">.</span><span class="n">getPairwiseDistances</span><span class="p">()</span>
</pre></div>
</div>
<p>or as a table for display / saving</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">dist_calc</span><span class="o">.</span><span class="n">Dists</span><span class="p">[:</span><span class="mi">4</span><span class="p">,:</span><span class="mi">4</span><span class="p">]</span> <span class="c"># truncated to fit screens</span>
<span class="go">Pairwise Distances</span>
<span class="go">============================================</span>
<span class="go">Seq1 \ Seq2    Galago    HowlerMon    Rhesus</span>
<span class="go">--------------------------------------------</span>
<span class="go">     Galago         *       0.2157    0.1962</span>
<span class="go">  HowlerMon    0.2157            *    0.0736</span>
<span class="go">     Rhesus    0.1962       0.0736         *</span>
<span class="go">  Orangutan    0.1944       0.0719    0.0411</span>
<span class="go">--------------------------------------------</span>
</pre></div>
</div>
<p>Other statistics are also available, such the as the standard errors of the estimates.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">dist_calc</span><span class="o">.</span><span class="n">StdErr</span><span class="p">[:</span><span class="mi">4</span><span class="p">,:</span><span class="mi">4</span><span class="p">]</span> <span class="c"># truncated to fit screens</span>
<span class="go">Standard Error of Pairwise Distances</span>
<span class="go">============================================</span>
<span class="go">Seq1 \ Seq2    Galago    HowlerMon    Rhesus</span>
<span class="go">--------------------------------------------</span>
<span class="go">     Galago         *       0.0103    0.0096</span>
<span class="go">  HowlerMon    0.0103            *    0.0054</span>
<span class="go">     Rhesus    0.0096       0.0054         *</span>
<span class="go">  Orangutan    0.0095       0.0053    0.0039</span>
<span class="go">--------------------------------------------</span>
</pre></div>
</div>
</div>
<div class="section" id="more-general-estimation-of-pairwise-distances">
<h4>More general estimation of pairwise distances<a class="headerlink" href="#more-general-estimation-of-pairwise-distances" title="Permalink to this headline">¶</a></h4>
<p>The standard cogent likelihood function can also be used to estimate distances. Because these require numerical optimisation they can be significantly slower than the fast estimation approach above.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.phylo</span> <span class="kn">import</span> <span class="n">distance</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.evolve.models</span> <span class="kn">import</span> <span class="n">F81</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="s">&#39;data/primate_brca1.fasta&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">distance</span><span class="o">.</span><span class="n">EstimateDistances</span><span class="p">(</span><span class="n">aln</span><span class="p">,</span> <span class="n">submodel</span><span class="o">=</span><span class="n">F81</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>The example above will use the F81 nucleotide substitution model and run the <tt class="docutils literal"><span class="pre">distance.EstimateDistances()</span></tt> method with the default options for the optimiser. To configure the optimiser a dictionary of optimisation options can be passed onto the <tt class="docutils literal"><span class="pre">run</span></tt> command. The example below configures the <tt class="docutils literal"><span class="pre">Powell</span></tt> optimiser to run a maximum of 10000 evaluations, with a maximum of 5 restarts (a total of 5 x 10000 = 50000 evaluations).</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">dist_opt_args</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">max_restarts</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">max_evaluations</span><span class="o">=</span><span class="mi">10000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">dist_opt_args</span><span class="o">=</span><span class="n">dist_opt_args</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">d</span>
<span class="go">============================================================================================</span>
<span class="go">Seq1 \ Seq2    Galago    HowlerMon    Rhesus    Orangutan    Gorilla     Human    Chimpanzee</span>
<span class="go">--------------------------------------------------------------------------------------------</span>
<span class="go">     Galago         *       0.2112    0.1930       0.1915     0.1891    0.1934        0.1892</span>
<span class="go">  HowlerMon    0.2112            *    0.0729       0.0713     0.0693    0.0729        0.0697</span>
<span class="go">     Rhesus    0.1930       0.0729         *       0.0410     0.0391    0.0421        0.0395</span>
<span class="go">  Orangutan    0.1915       0.0713    0.0410            *     0.0136    0.0173        0.0140</span>
<span class="go">    Gorilla    0.1891       0.0693    0.0391       0.0136          *    0.0086        0.0054</span>
<span class="go">      Human    0.1934       0.0729    0.0421       0.0173     0.0086         *        0.0089</span>
<span class="go"> Chimpanzee    0.1892       0.0697    0.0395       0.0140     0.0054    0.0089             *</span>
<span class="go">--------------------------------------------------------------------------------------------</span>
</pre></div>
</div>
</div>
<div class="section" id="building-a-phylogenetic-tree-from-pairwise-distances">
<h4>Building A Phylogenetic Tree From Pairwise Distances<a class="headerlink" href="#building-a-phylogenetic-tree-from-pairwise-distances" title="Permalink to this headline">¶</a></h4>
<p>Phylogenetic Trees can be built by using the neighbour joining algorithm by providing a dictionary of pairwise distances. This dictionary can be obtained either from the output of <tt class="docutils literal"><span class="pre">distance.EstimateDistances()</span></tt></p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.phylo</span> <span class="kn">import</span> <span class="n">nj</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">njtree</span> <span class="o">=</span> <span class="n">nj</span><span class="o">.</span><span class="n">nj</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">getPairwiseDistances</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">njtree</span> <span class="o">=</span> <span class="n">njtree</span><span class="o">.</span><span class="n">balanced</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">njtree</span><span class="o">.</span><span class="n">asciiArt</span><span class="p">()</span>
<span class="go">                    /-Rhesus</span>
<span class="go">          /edge.1--|</span>
<span class="go">         |         |          /-HowlerMon</span>
<span class="go">         |          \edge.0--|</span>
<span class="go">         |                    \-Galago</span>
<span class="go">-root----|</span>
<span class="go">         |--Orangutan</span>
<span class="go">         |</span>
<span class="go">         |          /-Human</span>
<span class="go">          \edge.2--|</span>
<span class="go">                   |          /-Gorilla</span>
<span class="go">                    \edge.3--|</span>
<span class="go">                              \-Chimpanzee</span>
</pre></div>
</div>
<p>Or created manually as shown below.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">dists</span> <span class="o">=</span> <span class="p">{(</span><span class="s">&#39;a&#39;</span><span class="p">,</span> <span class="s">&#39;b&#39;</span><span class="p">):</span> <span class="mf">2.7</span><span class="p">,</span> <span class="p">(</span><span class="s">&#39;c&#39;</span><span class="p">,</span> <span class="s">&#39;b&#39;</span><span class="p">):</span> <span class="mf">2.33</span><span class="p">,</span> <span class="p">(</span><span class="s">&#39;c&#39;</span><span class="p">,</span> <span class="s">&#39;a&#39;</span><span class="p">):</span> <span class="mf">0.73</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">njtree2</span> <span class="o">=</span> <span class="n">nj</span><span class="o">.</span><span class="n">nj</span><span class="p">(</span><span class="n">dists</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">njtree2</span><span class="o">.</span><span class="n">asciiArt</span><span class="p">()</span>
<span class="go">          /-a</span>
<span class="go">         |</span>
<span class="go">-root----|--b</span>
<span class="go">         |</span>
<span class="go">          \-c</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="by-least-squares">
<h3>By least-squares<a class="headerlink" href="#by-least-squares" title="Permalink to this headline">¶</a></h3>
<p>We illustrate the phylogeny reconstruction by least-squares using the F81 substitution model. We use the advanced-stepwise addition algorithm to search tree space. Here <tt class="docutils literal"><span class="pre">a</span></tt> is the number of taxa to exhaustively evaluate all possible phylogenies for. Successive taxa will are added to the top <tt class="docutils literal"><span class="pre">k</span></tt> trees (measured by the least-squares metric) and <tt class="docutils literal"><span class="pre">k</span></tt> trees are kept at each iteration.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">cPickle</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.phylo.least_squares</span> <span class="kn">import</span> <span class="n">WLS</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dists</span> <span class="o">=</span> <span class="n">cPickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s">&#39;data/dists_for_phylo.pickle&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ls</span> <span class="o">=</span> <span class="n">WLS</span><span class="p">(</span><span class="n">dists</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stat</span><span class="p">,</span> <span class="n">tree</span> <span class="o">=</span> <span class="n">ls</span><span class="o">.</span><span class="n">trex</span><span class="p">(</span><span class="n">a</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">show_progress</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
</pre></div>
</div>
<p>Other optional arguments that can be passed to the <tt class="docutils literal"><span class="pre">trex</span></tt> method are: <tt class="docutils literal"><span class="pre">return_all</span></tt>, whether the <tt class="docutils literal"><span class="pre">k</span></tt> best trees at the final step are returned as a <tt class="docutils literal"><span class="pre">ScoredTreeCollection</span></tt> object; <tt class="docutils literal"><span class="pre">order</span></tt>, a series of tip names whose order defines the sequence in which tips will be added during tree building (this allows the user to randomise the input order).</p>
</div>
<div class="section" id="by-ml">
<h3>By ML<a class="headerlink" href="#by-ml" title="Permalink to this headline">¶</a></h3>
<p>We illustrate the phylogeny reconstruction using maximum-likelihood using the F81 substitution model. We use the advanced-stepwise addition algorithm to search tree space, setting</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.phylo.maximum_likelihood</span> <span class="kn">import</span> <span class="n">ML</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.evolve.models</span> <span class="kn">import</span> <span class="n">F81</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="s">&#39;data/primate_brca1.fasta&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ml</span> <span class="o">=</span> <span class="n">ML</span><span class="p">(</span><span class="n">F81</span><span class="p">(),</span> <span class="n">aln</span><span class="p">)</span>
</pre></div>
</div>
<p>The <tt class="docutils literal"><span class="pre">ML</span></tt> object also has the <tt class="docutils literal"><span class="pre">trex</span></tt> method and this can be used in the same way as for above, i.e. <tt class="docutils literal"><span class="pre">ml.trex()</span></tt>. We don&#8217;t do that here because this is a very slow method for phylogenetic reconstruction.</p>
</div>
</div>
<div class="section" id="building-phylogenies-with-3rd-party-apps-such-as-fasttree-or-raxml">
<h2>Building phylogenies with 3rd-party apps such as FastTree or RAxML<a class="headerlink" href="#building-phylogenies-with-3rd-party-apps-such-as-fasttree-or-raxml" title="Permalink to this headline">¶</a></h2>
<p>A thorough description is <a class="reference internal" href="../examples/phylogeny_app_controllers.html#appcontroller-phylogeny"><em>Using phylogeny application controllers to construct phylogenetic trees from alignments</em></a>.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div class="news">
<table id="feed"><tr><td><h3><a href="http://pycogent.wordpress.com/">PyCogent News and Announcements</a></h3></td>
</tr></table></div>

    
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Building phylogenies</a><ul>
<li><a class="reference internal" href="#built-in-phylogenetic-reconstruction">Built-in Phylogenetic reconstruction</a><ul>
<li><a class="reference internal" href="#by-distance-method">By distance method</a><ul>
<li><a class="reference internal" href="#fast-pairwise-distance-estimation">Fast pairwise distance estimation</a></li>
<li><a class="reference internal" href="#more-general-estimation-of-pairwise-distances">More general estimation of pairwise distances</a></li>
<li><a class="reference internal" href="#building-a-phylogenetic-tree-from-pairwise-distances">Building A Phylogenetic Tree From Pairwise Distances</a></li>
</ul>
</li>
<li><a class="reference internal" href="#by-least-squares">By least-squares</a></li>
<li><a class="reference internal" href="#by-ml">By ML</a></li>
</ul>
</li>
<li><a class="reference internal" href="#building-phylogenies-with-3rd-party-apps-such-as-fasttree-or-raxml">Building phylogenies with 3rd-party apps such as FastTree or RAxML</a></li>
</ul>
</li>
</ul>



  <h4>Previous topic</h4>
  <p class="topless"><a href="building_a_tree_of_life.html"
                        title="previous chapter">Building a tree of life</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="using_likelihood_to_perform_evolutionary_analyses.html"
                        title="next chapter">Evolutionary analysis using likelihood</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/cookbook/building_phylogenies.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="using_likelihood_to_perform_evolutionary_analyses.html" title="Evolutionary analysis using likelihood"
             >next</a> |</li>
        <li class="right" >
          <a href="building_a_tree_of_life.html" title="Building a tree of life"
             >previous</a> |</li>
        <li><a href="../index.html">PyCogent 1.5.3 documentation</a> &raquo;</li>
          <li><a href="index.html" >PyCogent Cookbook</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2009, PyCogent Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>