

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Loading nucleotide, protein sequences &mdash; PyCogent 1.5.3 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.5.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PyCogent 1.5.3 documentation" href="../index.html" />
    <link rel="up" title="Manipulating biological data" href="manipulating_biological_data.html" />
    <link rel="next" title="Using the MolType and Sequence objects" href="moltypesequence.html" />
    <link rel="prev" title="Manipulating biological data" href="manipulating_biological_data.html" />     
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRQZ4X-jPDmWcshBrz2j7-fJvuUABRTGWmdiw2G89JpgztGlFGG8hDxRAw"></script>\
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="moltypesequence.html" title="Using the MolType and Sequence objects"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="manipulating_biological_data.html" title="Manipulating biological data"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">PyCogent 1.5.3 documentation</a> &raquo;</li>
          <li><a href="index.html" >PyCogent Cookbook</a> &raquo;</li>
          <li><a href="manipulating_biological_data.html" accesskey="U">Manipulating biological data</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="loading-nucleotide-protein-sequences">
<span id="load-seqs"></span><h1>Loading nucleotide, protein sequences<a class="headerlink" href="#loading-nucleotide-protein-sequences" title="Permalink to this headline">¶</a></h1>
<div class="section" id="loadseqs-from-a-file">
<h2><tt class="docutils literal"><span class="pre">LoadSeqs</span></tt> from a file<a class="headerlink" href="#loadseqs-from-a-file" title="Permalink to this headline">¶</a></h2>
<div class="section" id="as-an-alignment">
<h3>As an alignment<a class="headerlink" href="#as-an-alignment" title="Permalink to this headline">¶</a></h3>
<p>The function <tt class="docutils literal"><span class="pre">LoadSeqs()</span></tt> creates either a sequence collection or an alignment depending on the keyword argument <tt class="docutils literal"><span class="pre">aligned</span></tt> (the default is <tt class="docutils literal"><span class="pre">True</span></tt>).</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="s">&#39;data/long_testseqs.fasta&#39;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="n">DNA</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="go">&lt;class &#39;cogent.core.alignment.Alignment&#39;&gt;</span>
</pre></div>
</div>
<p>This example and some of the following use the <a class="reference download internal" href="../_downloads/long_testseqs.fasta"><tt class="xref download docutils literal"><span class="pre">long_testseqs.fasta</span></tt></a> file.</p>
</div>
<div class="section" id="as-a-sequence-collection-unaligned">
<h3>As a sequence collection (unaligned)<a class="headerlink" href="#as-a-sequence-collection-unaligned" title="Permalink to this headline">¶</a></h3>
<p>Setting the <tt class="docutils literal"><span class="pre">LoadSeqs()</span></tt> function keyword argument <tt class="docutils literal"><span class="pre">aligned=False</span></tt> returns a sequence collection.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="s">&#39;data/long_testseqs.fasta&#39;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="n">DNA</span><span class="p">,</span> <span class="n">aligned</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="nb">type</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="go">&lt;class &#39;cogent.core.alignment.SequenceCollection&#39;&gt;</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">An alignment can be sliced, but a <tt class="docutils literal"><span class="pre">SequenceCollection</span></tt> can not.</p>
</div>
</div>
<div class="section" id="specifying-the-file-format">
<h3>Specifying the file format<a class="headerlink" href="#specifying-the-file-format" title="Permalink to this headline">¶</a></h3>
<p><tt class="docutils literal"><span class="pre">LoadSeqs()</span></tt> uses the filename suffix to infer the file format. This can be overridden using the <tt class="docutils literal"><span class="pre">format</span></tt> argument.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="s">&#39;data/long_testseqs.fasta&#39;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="n">DNA</span><span class="p">,</span>
<span class="gp">... </span>                 <span class="n">format</span><span class="o">=</span><span class="s">&#39;fasta&#39;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span>
<span class="go">5 x 2532 dna alignment: Human[TGTGGCACAAA...</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="loadseqs-from-a-series-of-strings">
<h2><tt class="docutils literal"><span class="pre">LoadSeqs</span></tt> from a series of strings<a class="headerlink" href="#loadseqs-from-a-series-of-strings" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;&gt;seq1&#39;</span><span class="p">,</span><span class="s">&#39;AATCG-A&#39;</span><span class="p">,</span><span class="s">&#39;&gt;seq2&#39;</span><span class="p">,</span><span class="s">&#39;AATCGGA&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs_loaded</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">seqs_loaded</span>
<span class="go">&gt;seq1</span>
<span class="go">AATCG-A</span>
<span class="go">&gt;seq2</span>
<span class="go">AATCGGA</span>
</pre></div>
</div>
</div>
<div class="section" id="loadseqs-from-a-dict-of-strings">
<h2><tt class="docutils literal"><span class="pre">LoadSeqs</span></tt> from a dict of strings<a class="headerlink" href="#loadseqs-from-a-dict-of-strings" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;seq1&#39;</span><span class="p">:</span> <span class="s">&#39;AATCG-A&#39;</span><span class="p">,</span><span class="s">&#39;seq2&#39;</span><span class="p">:</span> <span class="s">&#39;AATCGGA&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs_loaded</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">seqs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="specifying-the-sequence-molecular-type">
<h2>Specifying the sequence molecular type<a class="headerlink" href="#specifying-the-sequence-molecular-type" title="Permalink to this headline">¶</a></h2>
<p>Simple case of loading a <tt class="docutils literal"><span class="pre">list</span></tt> of aligned amino acid sequences in FASTA format, with and without molecule type specification. When the <tt class="docutils literal"><span class="pre">MolType</span></tt> is not specified it defaults to ASCII.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">PROTEIN</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protein_seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;&gt;seq1&#39;</span><span class="p">,</span><span class="s">&#39;DEKQL-RG&#39;</span><span class="p">,</span><span class="s">&#39;&gt;seq2&#39;</span><span class="p">,</span><span class="s">&#39;DDK--SRG&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">proteins_loaded</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">protein_seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">proteins_loaded</span><span class="o">.</span><span class="n">MolType</span>
<span class="go">MolType((&#39;a&#39;, &#39;b&#39;, &#39;c&#39;, &#39;d&#39;, &#39;e&#39;, ...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">proteins_loaded</span>
<span class="go">&gt;seq1</span>
<span class="go">DEKQL-RG</span>
<span class="go">&gt;seq2</span>
<span class="go">DDK--SRG</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">proteins_loaded</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">protein_seqs</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="n">PROTEIN</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">proteins_loaded</span>
<span class="go">&gt;seq1</span>
<span class="go">DEKQL-RG</span>
<span class="go">&gt;seq2</span>
<span class="go">DDK--SRG</span>
</pre></div>
</div>
</div>
<div class="section" id="stripping-label-characters-on-loading">
<h2>Stripping label characters on loading<a class="headerlink" href="#stripping-label-characters-on-loading" title="Permalink to this headline">¶</a></h2>
<p>Load a list of aligned nucleotide sequences, while specifying the DNA molecule type and stripping the comments from the label. In this example, stripping is accomplished by passing a function that removes everything after the first whitespace to the <tt class="docutils literal"><span class="pre">label_to_name</span></tt> parameter.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">DNA_seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;&gt;sample1 Mus musculus&#39;</span><span class="p">,</span><span class="s">&#39;AACCTGC--C&#39;</span><span class="p">,</span><span class="s">&#39;&gt;sample2 Gallus gallus&#39;</span><span class="p">,</span><span class="s">&#39;AAC-TGCAAC&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loaded_seqs</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">DNA_seqs</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="n">DNA</span><span class="p">,</span> <span class="n">label_to_name</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">loaded_seqs</span>
<span class="go">&gt;sample1</span>
<span class="go">AACCTGC--C</span>
<span class="go">&gt;sample2</span>
<span class="go">AAC-TGCAAC</span>
</pre></div>
</div>
</div>
<div class="section" id="using-alternative-constructors-for-the-alignment-object">
<h2>Using alternative constructors for the <cite>Alignment</cite> object<a class="headerlink" href="#using-alternative-constructors-for-the-alignment-object" title="Permalink to this headline">¶</a></h2>
<p>An example of using an alternative constructor is given below. A constructor is passed to the aligned parameter in lieu of <tt class="docutils literal"><span class="pre">True</span></tt> or <tt class="docutils literal"><span class="pre">False</span></tt>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent</span> <span class="kn">import</span> <span class="n">LoadSeqs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.core.alignment</span> <span class="kn">import</span> <span class="n">DenseAlignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;&gt;seq1&#39;</span><span class="p">,</span><span class="s">&#39;AATCG-A&#39;</span><span class="p">,</span><span class="s">&#39;&gt;seq2&#39;</span><span class="p">,</span><span class="s">&#39;AATCGGA&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs_loaded</span> <span class="o">=</span> <span class="n">LoadSeqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">seqs</span><span class="p">,</span><span class="n">aligned</span><span class="o">=</span><span class="n">DenseAlignment</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">seqs_loaded</span>
<span class="go">&gt;seq1</span>
<span class="go">AATCG-A</span>
<span class="go">&gt;seq2</span>
<span class="go">AATCGGA</span>
</pre></div>
</div>
</div>
<div class="section" id="loading-sequences-using-format-parsers">
<h2>Loading sequences using format parsers<a class="headerlink" href="#loading-sequences-using-format-parsers" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal"><span class="pre">LoadSeqs</span></tt> is just a convenience interface to format parsers. It can sometimes be more effective to use the parsers directly, say when you don&#8217;t want to load everything into memory.</p>
<div class="section" id="loading-fasta-sequences-from-an-open-file-or-list-of-lines">
<h3>Loading FASTA sequences from an open file or list of lines<a class="headerlink" href="#loading-fasta-sequences-from-an-open-file-or-list-of-lines" title="Permalink to this headline">¶</a></h3>
<p>To load FASTA formatted sequences directly, you can use the <tt class="docutils literal"><span class="pre">MinimalFastaParser</span></tt>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This returns the sequences as strings.</p>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.parse.fasta</span> <span class="kn">import</span> <span class="n">MinimalFastaParser</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="s">&#39;data/long_testseqs.fasta&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[(</span><span class="n">name</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">MinimalFastaParser</span><span class="p">(</span><span class="n">f</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">seqs</span>
<span class="go">[(&#39;Human&#39;, &#39;TGTGGCACAAATAC...</span>
</pre></div>
</div>
</div>
<div class="section" id="handling-overloaded-fasta-sequence-labels">
<h3>Handling overloaded FASTA sequence labels<a class="headerlink" href="#handling-overloaded-fasta-sequence-labels" title="Permalink to this headline">¶</a></h3>
<p>The FASTA label field is frequently overloaded, with different information fields present in the field and separated by some delimiter. This can be flexibly addressed using the <tt class="docutils literal"><span class="pre">LabelParser</span></tt>. By creating a custom label parser, we can decided which part we use as the sequence name. We show how convert a field into something specific.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.parse.fasta</span> <span class="kn">import</span> <span class="n">LabelParser</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">latin_to_common</span><span class="p">(</span><span class="n">latin</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="p">{</span><span class="s">&#39;Homo sapiens&#39;</span><span class="p">:</span> <span class="s">&#39;human&#39;</span><span class="p">,</span>
<span class="gp">... </span>            <span class="s">&#39;Pan troglodtyes&#39;</span><span class="p">:</span> <span class="s">&#39;chimp&#39;</span><span class="p">}[</span><span class="n">latin</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">label_parser</span> <span class="o">=</span> <span class="n">LabelParser</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%(species)s</span><span class="s">&quot;</span><span class="p">,</span>
<span class="gp">... </span>            <span class="p">[[</span><span class="mi">1</span><span class="p">,</span> <span class="s">&quot;species&quot;</span><span class="p">,</span> <span class="n">latin_to_common</span><span class="p">]],</span> <span class="n">split_with</span><span class="o">=</span><span class="s">&#39;:&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="s">&quot;&gt;abcd:Homo sapiens:misc&quot;</span><span class="p">,</span> <span class="s">&quot;&gt;abcd:Pan troglodtyes:misc&quot;</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">label</span> <span class="o">=</span> <span class="n">label_parser</span><span class="p">(</span><span class="n">label</span><span class="p">)</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">label</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="n">label</span><span class="p">)</span>
<span class="go">human &lt;class &#39;cogent.parse.fasta.RichLabel&#39;&gt;</span>
<span class="go">chimp &lt;class &#39;cogent.parse.fasta.RichLabel&#39;&gt;</span>
</pre></div>
</div>
<p>The <tt class="docutils literal"><span class="pre">RichLabel</span></tt> objects have an <tt class="docutils literal"><span class="pre">Info</span></tt> object as an attribute, allowing specific reference to all the specified label fields.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent.parse.fasta</span> <span class="kn">import</span> <span class="n">MinimalFastaParser</span><span class="p">,</span> <span class="n">LabelParser</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fasta_data</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;&gt;gi|10047090|ref|NP_055147.1| small muscle protein, X-linked [Homo sapiens]&#39;</span><span class="p">,</span>
<span class="gp">... </span> <span class="s">&#39;MNMSKQPVSNVRAIQANINIPMGAFRPGAGQPPRRKECTPEVEEGVPPTSDEEKKPIPGAKKLPGPAVNL&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="s">&#39;SEIQNIKSELKYVPKAEQ&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="s">&#39;&gt;gi|10047092|ref|NP_037391.1| neuronal protein [Homo sapiens]&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="s">&#39;MANRGPSYGLSREVQEKIEQKYDADLENKLVDWIILQCAEDIEHPPPGRAHFQKWLMDGTVLCKLINSLY&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="s">&#39;PPGQEPIPKISESKMAFKQMEQISQFLKAAETYGVRTTDIFQTVDLWEGKDMAAVQRTLMALGSVAVTKD&#39;</span><span class="p">]</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">label_to_name</span> <span class="o">=</span> <span class="n">LabelParser</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%(ref)s</span><span class="s">&quot;</span><span class="p">,</span>
<span class="gp">... </span>                             <span class="p">[[</span><span class="mi">1</span><span class="p">,</span><span class="s">&quot;gi&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
<span class="gp">... </span>                              <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="s">&quot;ref&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
<span class="gp">... </span>                              <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="s">&quot;description&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span>
<span class="gp">... </span>                              <span class="n">split_with</span><span class="o">=</span><span class="s">&quot;|&quot;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">MinimalFastaParser</span><span class="p">(</span><span class="n">fasta_data</span><span class="p">,</span> <span class="n">label_to_name</span><span class="o">=</span><span class="n">label_to_name</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">name</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">name</span><span class="o">.</span><span class="n">Info</span><span class="o">.</span><span class="n">gi</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">name</span><span class="o">.</span><span class="n">Info</span><span class="o">.</span><span class="n">description</span>
<span class="go">NP_055147.1</span>
<span class="go">10047090</span>
<span class="go"> small muscle protein, X-linked [Homo sapiens]</span>
<span class="go">NP_037391.1</span>
<span class="go">10047092</span>
<span class="go"> neuronal protein [Homo sapiens]</span>
</pre></div>
</div>
</div>
<div class="section" id="loading-dna-sequences-from-a-genbank-file">
<h3>Loading DNA sequences from a GenBank file<a class="headerlink" href="#loading-dna-sequences-from-a-genbank-file" title="Permalink to this headline">¶</a></h3>
<p><em>To be written.</em></p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div class="news">
<table id="feed"><tr><td><h3><a href="http://pycogent.wordpress.com/">PyCogent News and Announcements</a></h3></td>
</tr></table></div>

    
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Loading nucleotide, protein sequences</a><ul>
<li><a class="reference internal" href="#loadseqs-from-a-file"><tt class="docutils literal"><span class="pre">LoadSeqs</span></tt> from a file</a><ul>
<li><a class="reference internal" href="#as-an-alignment">As an alignment</a></li>
<li><a class="reference internal" href="#as-a-sequence-collection-unaligned">As a sequence collection (unaligned)</a></li>
<li><a class="reference internal" href="#specifying-the-file-format">Specifying the file format</a></li>
</ul>
</li>
<li><a class="reference internal" href="#loadseqs-from-a-series-of-strings"><tt class="docutils literal"><span class="pre">LoadSeqs</span></tt> from a series of strings</a></li>
<li><a class="reference internal" href="#loadseqs-from-a-dict-of-strings"><tt class="docutils literal"><span class="pre">LoadSeqs</span></tt> from a dict of strings</a></li>
<li><a class="reference internal" href="#specifying-the-sequence-molecular-type">Specifying the sequence molecular type</a></li>
<li><a class="reference internal" href="#stripping-label-characters-on-loading">Stripping label characters on loading</a></li>
<li><a class="reference internal" href="#using-alternative-constructors-for-the-alignment-object">Using alternative constructors for the <cite>Alignment</cite> object</a></li>
<li><a class="reference internal" href="#loading-sequences-using-format-parsers">Loading sequences using format parsers</a><ul>
<li><a class="reference internal" href="#loading-fasta-sequences-from-an-open-file-or-list-of-lines">Loading FASTA sequences from an open file or list of lines</a></li>
<li><a class="reference internal" href="#handling-overloaded-fasta-sequence-labels">Handling overloaded FASTA sequence labels</a></li>
<li><a class="reference internal" href="#loading-dna-sequences-from-a-genbank-file">Loading DNA sequences from a GenBank file</a></li>
</ul>
</li>
</ul>
</li>
</ul>



  <h4>Previous topic</h4>
  <p class="topless"><a href="manipulating_biological_data.html"
                        title="previous chapter">Manipulating biological data</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="moltypesequence.html"
                        title="next chapter">Using the <tt class="docutils literal docutils literal"><span class="pre">MolType</span></tt> and <tt class="docutils literal docutils literal"><span class="pre">Sequence</span></tt> objects</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/cookbook/loading_sequences.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="moltypesequence.html" title="Using the MolType and Sequence objects"
             >next</a> |</li>
        <li class="right" >
          <a href="manipulating_biological_data.html" title="Manipulating biological data"
             >previous</a> |</li>
        <li><a href="../index.html">PyCogent 1.5.3 documentation</a> &raquo;</li>
          <li><a href="index.html" >PyCogent Cookbook</a> &raquo;</li>
          <li><a href="manipulating_biological_data.html" >Manipulating biological data</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2009, PyCogent Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>